

<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Gene reads script &#8212; SATAY DATA ANALYSIS</title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha384-KA6wR/X5RY4zFAHpv/CnoG2UW1uogYfdnP67Uv7eULvTveboZJg0qUpmJZb5VqzN" crossorigin="anonymous">
    <link href="_static/css/index.css" rel="stylesheet">
    <link rel="stylesheet" href="_static/sphinx-book-theme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/jupyter-sphinx.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/sphinx-book-theme.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/mystnb.js"></script>
    <script src="_static/sphinx-book-theme.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script async="async" src="_static/thebelab.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="_static/Greg.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Statistics per gene" href="statistics_pergene.html" />
    <link rel="prev" title="Notebooks written entirely in markdown" href="MysT_Notebooks.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="docsearch:language" content="en">



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="index.html">
  
  <img src="_static/Greg.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">SATAY DATA ANALYSIS</h1>
  
</a>
</div>

<form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>

<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  
  <ul class="nav sidenav_l1">
  <li class="">
    <a href="Intro.html">1. To reproduce an example of  Jupyter book, and you are a Windows user:</a>
  </li>
<li class="navbar-special">
<p class="margin-caption">Get started</p>
</li>
  <li class="">
    <a href="Intro/satay_analysis_notes.html">2. All what you need to know from SATAY</a>
  </li>
  <li class="">
    <a href="Intro/data-analysis-workflow.html">3. Data analysis steps</a>
  </li>
  <li class="">
    <a href="MysT_Notebooks.html">4. Notebooks written entirely in markdown</a>
  </li>
<li class="navbar-special">
<p class="margin-caption">Postprocessing scripts</p>
</li>
  <li class="active">
    <a href="">5. Gene reads script</a>
  </li>
  <li class="">
    <a href="statistics_pergene.html">6. Statistics per gene</a>
  </li>
</ul>
</nav>

 <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse" data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu" aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation" title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        <div class="dropdown-buttons-trigger">
            <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i class="fas fa-download"></i></button>

            
            <div class="dropdown-buttons">
                <!-- ipynb file if we had a myst markdown file -->
                <a class="dropdown-buttons" href="_sources/gene_reads.ipynb"><button type="button" class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip" data-placement="left">.ipynb</button></a>
                <!-- Download raw file -->
                <a class="dropdown-buttons" href="_sources/gene_reads.md"><button type="button" class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip" data-placement="left">.md</button></a>
                <!-- Download PDF via print -->
                <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF" onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
            </div>
            
        </div>

        <!-- Source interaction buttons -->
        

        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="bottom" onclick="toggleFullScreen()" title="Fullscreen mode"><i class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->
        
        <div class="dropdown-buttons-trigger">
            <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
            <div class="dropdown-buttons">
                
                
                <button type="button" class="btn btn-secondary topbarbtn thebelab-launch-button" onclick="initThebelab()" title="Launch Thebelab" data-toggle="tooltip" data-placement="left"><i class="fas fa-rocket"></i><span style="margin-left: .4em;">ThebeLab</span></button>
                
            </div>
        </div>
        
    </div>
    <div class="d-none d-md-block col-md-2 bd-toc show">
<div class="tocsection onthispage pt-5 pb-3">
    <i class="fas fa-list"></i> On this page
</div>

<nav id="bd-toc-nav">
    <ul class="nav section-nav flex-column">
    
        <li class="nav-item toc-entry toc-h2">
            <a href="#introduction" class="nav-link">Introduction</a>
        </li>
    
        <li class="nav-item toc-entry toc-h2">
            <a href="#the-python-script" class="nav-link">The python script</a><ul class="nav section-nav flex-column">
                
        <li class="nav-item toc-entry toc-h3">
            <a href="#input" class="nav-link">Input</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#loading-files" class="nav-link">Loading files</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#get-start-and-end-position-of-the-gene-or-region" class="nav-link">Get start and end position of the gene or region</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#get-the-reads" class="nav-link">Get the reads</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#consider-double-insertions" class="nav-link">Consider double insertions</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#create-list-of-all-locations-in-the-gene" class="nav-link">Create list of all locations in the gene</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#statistics-and-prepare-for-plotting" class="nav-link">Statistics and prepare for plotting</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#binning" class="nav-link">Binning</a>
        </li>
    
        <li class="nav-item toc-entry toc-h3">
            <a href="#plotting" class="nav-link">Plotting</a>
        </li>
    
            </ul>
        </li>
    
        <li class="nav-item toc-entry toc-h2">
            <a href="#interpretation" class="nav-link">Interpretation</a>
        </li>
    
        <li class="nav-item toc-entry toc-h2">
            <a href="#bibliography" class="nav-link">Bibliography</a>
        </li>
    
    </ul>
</nav>


    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="gene-reads-script">
<h1>Gene reads script<a class="headerlink" href="#gene-reads-script" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>To get an indication of possible essential genes, the number of transposon insertions and amount of reads per transposon is plotted along the length of the gene. If a gene consists of a large area without any insertions or with little reads, this might indicate an essential (region of a) gene.
However, it sometimes might happen that non-essential genes have no transposons. This can mulitple reasons, most notable which are:</p>
<ul class="simple">
<li><p>Sequencing reads that map to repeated sequences were discarded during alignment. If these repeated sequences occur in protein coding regions in the genome, then these regions might be removed during alignment and therefore these genes appear to be free of transposon, even though these genes might not be essential.</p></li>
<li><p>Annotated dubious ORFs might overlap with essential genes. So the dubious ORFs are not essential, but have no insertions because of the overlap.</p></li>
<li><p>Sometimes genes might not be essential in normal growth conditions (and hence are not annotated as such), but in particular lab conditions some genes can become essential (for example genes involved in the metabolism of certain sugars).</p></li>
</ul>
<p>Also, sometimes transposons occur in essential genes. This can be because not the complete gene is essential, but only certain domains of the gene are essential. The non-essential domain can be truncated and still create a function protein with only the essential domains left. Also, sometimes the first and/or last few basepairs can be truncated while still leave a gene that can be transcribed and translated in a functional protein.</p>
<p>For a more thorough discussion, see the paper from Michel et.al. 2017.</p>
</div>
<div class="section" id="the-python-script">
<h2>The python script<a class="headerlink" href="#the-python-script" title="Permalink to this headline">¶</a></h2>
<p>This function creates a bar plot with the number of reads per transposon for regions in a gene. It also shows the exact locations of the insertions along the gene. The width of the bars are chosen such that each bar include a fixed amount of transposons. A maximum threshold is set to account for the situation where there are little or no transposons present for large portion of the gene.</p>
<p>Next to the bar plot, this function also creates a violin plot where a distribution is shown for the distance (in terms of basepairs) between subsequent insertions. This is compared with the distance between insertions in the entire chromosome the gene is in. If the median distance between insertions is significantly higher in the gene compared to the chromosome, then this might be an indication of an essential gene.</p>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>The function inputs either a gene name (<code class="docutils literal notranslate"><span class="pre">gene_name</span></code>, type=string) or a region (<code class="docutils literal notranslate"><span class="pre">region</span></code>, type=list) specified as a list with three entries (chromosome number as a roman numeral, start position of the region and the end position respectively).
The variable <code class="docutils literal notranslate"><span class="pre">gene_name</span></code> can be set to any gene name or <code class="docutils literal notranslate"><span class="pre">holocus</span></code> or <code class="docutils literal notranslate"><span class="pre">ho-locus</span></code>.
Next it requires the bed-file (<code class="docutils literal notranslate"><span class="pre">bed_file</span></code>, type=string) which is created by the Matlab code provided from the Kornman lab [https://sites.google.com/site/satayusers/complete-protocol/bioinformatics-analysis/matlab-script].
Finally, the figure can be automatically saved (at a location specified in the beginning of the function) by setting <code class="docutils literal notranslate"><span class="pre">savefigure</span></code> to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<p>The custom build functions (stored in the ‘python modules’ folder on Github [https://github.com/Gregory94/LaanLab-SATAY-DataAnalysis/tree/master/python_modules]) that are required are:</p>
<ul class="simple">
<li><p>chromosome_and_gene_positions.gene_position</p></li>
<li><p>gene_names.gene_aliases</p></li>
<li><p>chromosome_names_in_files.chromosome_props_bedfile</p></li>
<li><p>statistics_perchromosome</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sb</span>

<span class="c1"># file_dirname = os.path.dirname(os.path.abspath(&#39;__file__&#39;))</span>
<span class="n">file_dirname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="s1">&#39;__file__&#39;</span><span class="p">)</span>
<span class="c1"># sys.path.insert(1,os.path.join(file_dirname,&#39;python_modules&#39;))</span>
<span class="c1"># os.chdir(&#39;mini_book/docs/Python_scripts/python_modules/&#39;) # with this path works in the local build</span>
<span class="kn">from</span> <span class="nn">Python_scripts.python_modules.chromosome_and_gene_positions</span> <span class="kn">import</span> <span class="n">gene_position</span>
<span class="kn">from</span> <span class="nn">Python_scripts.python_modules.gene_names</span> <span class="kn">import</span> <span class="n">gene_aliases</span>
<span class="kn">from</span> <span class="nn">Python_scripts.python_modules.chromosome_names_in_files</span> <span class="kn">import</span> <span class="n">chromosome_name_bedfile</span>
<span class="kn">import</span> <span class="nn">Python_scripts.python_modules.statistics_perchromosome</span> <span class="k">as</span> <span class="nn">statistics_perchromosome</span>


<span class="c1">#FUNCTION INPUTS</span>
<span class="n">gene_name</span> <span class="o">=</span> <span class="s1">&#39;bem1&#39;</span>
<span class="n">region</span><span class="o">=</span><span class="kc">None</span>

<span class="n">bed_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;satay_analysis_testdata&#39;</span><span class="p">,</span><span class="s1">&#39;Output_Processing&#39;</span><span class="p">,</span><span class="s1">&#39;Cerevisiae_WT2_Michel2017_trimmed1.bam.bed&#39;</span><span class="p">)</span><span class="c1">#CHANGE THIS TO ANY .BED FILE YOU WANT TO ANALYSE.</span>

<span class="nb">print</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;satay_analysis_testdata&#39;</span><span class="p">,</span><span class="s1">&#39;Output_Processing&#39;</span><span class="p">)))</span>
<span class="n">savefigure</span><span class="o">=</span><span class="kc">False</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="loading-files">
<h3>Loading files<a class="headerlink" href="#loading-files" title="Permalink to this headline">¶</a></h3>
<p>Next additional files are loaded. Change these to your local paths leading to the gff-file (for example downloaded from SGD [https://www.yeastgenome.org/] or get a copy from the docs folder on Github [https://github.com/Gregory94/LaanLab-SATAY-DataAnalysis]) and yeast_protein_Names file.
The <code class="docutils literal notranslate"><span class="pre">save_figure_path</span></code> directs to the location where the figures need to be stored.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># os.chdir(&#39;../&#39;)</span>
<span class="n">file_dirname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="s1">&#39;__file__&#39;</span><span class="p">)</span>

<span class="n">gff_file</span> <span class="o">=</span> <span class="s1">&#39;Python_scripts/Data_Files/Saccharomyces_cerevisiae.R64-1-1.99.gff3&#39;</span>
<span class="n">abs_gff_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span> <span class="n">gff_file</span><span class="p">)</span>
<span class="c1"># gff_file = os.path.join(file_dirname,&#39;Data_Files&#39;,&#39;Saccharomyces_cerevisiae.R64-1-1.99.gff3&#39;)</span>
<span class="n">gene_information_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;Python_scripts&#39;</span><span class="p">,</span><span class="s1">&#39;Data_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Yeast_Protein_Names.txt&#39;</span><span class="p">)</span>


<span class="k">if</span> <span class="n">savefigure</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
    <span class="n">save_figure_path</span> <span class="o">=</span> <span class="n">file_dirname</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="get-start-and-end-position-of-the-gene-or-region">
<h3>Get start and end position of the gene or region<a class="headerlink" href="#get-start-and-end-position-of-the-gene-or-region" title="Permalink to this headline">¶</a></h3>
<p>When a gene name is given, this part searches in the .gff file for the position and reading orientation of the gene.
However, often genes have multiple names and the name given by the user might be present as such in the .gff file.
Therefore this code searches for potential aliases of the gene name in the Yeast_Protein_Names.txt file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">gene_name</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="s1">&#39;HOLOCUS&#39;</span> <span class="ow">or</span> <span class="n">gene_name</span> <span class="o">==</span> <span class="s1">&#39;HO-LOCUS&#39;</span><span class="p">:</span>
    <span class="n">gene_pos</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IV&#39;</span><span class="p">,</span><span class="mi">46271</span><span class="p">,</span><span class="mi">48031</span><span class="p">]</span>
    <span class="n">gene_name</span> <span class="o">=</span> <span class="s1">&#39;HOlocus&#39;</span>

<span class="k">elif</span> <span class="n">gene_name</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">gene_pos_dict</span> <span class="o">=</span> <span class="n">gene_position</span><span class="p">(</span><span class="n">gff_file</span><span class="p">)</span> <span class="c1">#GET POSITION INFORMATION OF ALL GENES</span>
    
    <span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene_name</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="c1">#CAPITALIZE GENE NAME</span>
    <span class="k">if</span> <span class="n">gene_pos_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span> <span class="c1">#CHECK IF GENE_NAME EXISTS IN GENE_POS_DICT. IF NOT, CHECK IF ANY OF THE ALIASES EXISTS</span>
        <span class="n">gene_alias_dict</span> <span class="o">=</span> <span class="n">gene_aliases</span><span class="p">(</span><span class="n">gene_information_file</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">gene_alias_key</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">gene_alias_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">v</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">gene_pos_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_alias_key</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span> <span class="c1">#IF KEY DOES ALSO NOT EXISTS IN GENE_POS_DICT, CHECK IF MORE ALIASES EXISTS OF GENE_NAME</span>
            <span class="n">gene_alias_list</span> <span class="o">=</span> <span class="n">gene_alias_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_alias_key</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">gene_alias</span> <span class="ow">in</span> <span class="n">gene_alias_list</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gene_pos_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_alias</span><span class="p">)</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">gene_pos</span> <span class="o">=</span> <span class="n">gene_pos_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_alias</span><span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;The alias &#39;</span><span class="p">,</span><span class="n">gene_alias</span><span class="p">,</span> <span class="s1">&#39; is used for the gene &#39;</span><span class="p">,</span><span class="n">gene_name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_pos</span> <span class="o">=</span> <span class="n">gene_pos_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_alias_key</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;The alias &#39;</span><span class="p">,</span><span class="n">gene_alias_key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39; is used for the gene &#39;</span><span class="p">,</span><span class="n">gene_name</span><span class="p">)</span>
            
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gene_pos</span> <span class="o">=</span> <span class="n">gene_pos_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>

    <span class="n">gene_orien</span> <span class="o">=</span> <span class="n">gene_pos</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

<span class="k">elif</span> <span class="n">region</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">gene_pos</span> <span class="o">=</span> <span class="n">region</span>
    

<span class="n">gene_chr</span> <span class="o">=</span> <span class="n">gene_pos</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">gene_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">gene_end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene_pos</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="k">if</span> <span class="n">gene_name</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="s1">&#39; starts at basepair &#39;</span><span class="p">,</span><span class="n">gene_start</span><span class="p">,</span> <span class="s1">&#39; and ends at basepair &#39;</span><span class="p">,</span><span class="n">gene_end</span><span class="p">,</span> <span class="s1">&#39; in chromosome&#39;</span><span class="p">,</span><span class="n">gene_chr</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Selected region starts at basepair &#39;</span><span class="p">,</span><span class="n">gene_start</span><span class="p">,</span> <span class="s1">&#39; and ends at basepair &#39;</span><span class="p">,</span><span class="n">gene_end</span><span class="p">,</span> <span class="s1">&#39; in chromosome&#39;</span><span class="p">,</span><span class="n">gene_chr</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>The alias  YBR200W  is used for the gene  BEM1
BEM1  starts at basepair  620872  and ends at basepair  622527  in chromosome II
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="get-the-reads">
<h3>Get the reads<a class="headerlink" href="#get-the-reads" title="Permalink to this headline">¶</a></h3>
<p>Next, get the reads per transposon as stored in the bed file. For this the bed file is read and the lines are determined where in the bed file the insertions are for the chromosome where the gene is located in (using the chromosome_props_bedfile.py module). Within the chromosome, the insertion locations are searched that fall within the location of the gene.
the values for the reads that found in the bed file are determined by taking the value -100 and divided by 20. This is undo what is done in the Matlab code by Benoit for better visualization, but in this code we want to take the actual values and therefore the formula used by the Matlab code is reversed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">bed_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

<span class="n">chrom_start_line_dict</span><span class="p">,</span> <span class="n">chrom_end_line_dict</span><span class="o">=</span> <span class="n">chromosome_name_bedfile</span><span class="p">(</span><span class="n">lines</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>

<span class="n">insertion_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">read_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">[</span><span class="n">chrom_start_line_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">):</span><span class="n">chrom_end_line_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">)]:</span>
    <span class="n">line_list</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">gene_start</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line_list</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="n">gene_end</span><span class="p">:</span>
        <span class="n">insertion_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">line_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

        <span class="n">read_value</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">line_list</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span><span class="o">-</span><span class="mi">100</span><span class="p">)</span><span class="o">/</span><span class="mi">20</span> <span class="c1">#the matlab script by benoit takes the number of reads*20+100. This line makes this undone</span>
        <span class="n">read_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read_value</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="consider-double-insertions">
<h3>Consider double insertions<a class="headerlink" href="#consider-double-insertions" title="Permalink to this headline">¶</a></h3>
<p>Sometimes a single location has multiple insertions. This piece of code makes sure that the reads from all insertions at the same loation are accounted for.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">unique_insertion_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">duplicate_insertion_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">ins</span> <span class="ow">in</span> <span class="n">insertion_list</span><span class="p">:</span> <span class="c1">#FIND THE CHROMOSOME POSITION OF ALL DUPLICATED TRANSPOSON INSERTION SITES</span>
    <span class="k">if</span> <span class="n">ins</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">unique_insertion_list</span><span class="p">:</span>
        <span class="n">unique_insertion_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ins</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">duplicate_insertion_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ins</span><span class="p">)</span>
<span class="n">duplicate_insertion_list</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">duplicate_insertion_list</span><span class="p">)</span> <span class="c1">#ACCOUNT FOR THE SITUATION WHERE THERE ARE MORE THAN TWO INSERTIONS AT THE SAME LOCATION</span>

<span class="n">duplicate_index_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">dup</span> <span class="ow">in</span> <span class="n">duplicate_insertion_list</span><span class="p">:</span>
    <span class="n">insertion_arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">insertion_list</span><span class="p">)</span>
    <span class="n">duplicate_index_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">insertion_arr</span> <span class="o">==</span> <span class="n">dup</span><span class="p">))</span> <span class="c1">#GET ALL INDICES OF THE LIST OF TRANSPOSON INSERTIONS WHERE THE DUPLICATES ARE PRESENT. EACH INSERTION LOCATION GETS ITS OWN NUMPY ARRAY WITHIN THIS LIST</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">duplicate_index_list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">number_of_duplicates_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">insertion_list</span><span class="p">)</span> <span class="c1">#MAKE LIST OF ONES WITH SAME LENGTH AS INSERTION_LIST FOR STORING NUMBER OF DUPLICATES</span>
    <span class="n">delete_index</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">ind_arr</span> <span class="ow">in</span> <span class="n">duplicate_index_list</span><span class="p">:</span> <span class="c1">#LOOP OVER ALL INDICES OF DUPLICATES</span>
        <span class="n">ind_list</span> <span class="o">=</span> <span class="n">ind_arr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">ind_list_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">ind_list</span><span class="p">)</span> <span class="c1">#GET THE LAST INDEX OF THE DUPLICATES</span>
<span class="c1">#            print(&#39;Mulitple transposons found at &#39;,ind_list)</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">ind_list</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">ind</span> <span class="o">==</span> <span class="n">ind_list_max</span><span class="p">:</span>
                <span class="n">read_list</span><span class="p">[</span><span class="n">ind_list_max</span><span class="p">]</span> <span class="o">+=</span> <span class="n">read_list</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span> <span class="c1">#ADD UP THE READS TO THE LAST DUPLICATE</span>
                <span class="n">number_of_duplicates_list</span><span class="p">[</span><span class="n">ind_list_max</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ind_list</span><span class="p">)</span> <span class="c1">#UPDATE NUMBER OF DUPLICATES</span>
                <span class="n">delete_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>

    <span class="c1">#REVERSE LOOP OVER LIST FOR DELETING</span>
    <span class="k">for</span> <span class="n">del_ind</span> <span class="ow">in</span> <span class="nb">reversed</span><span class="p">(</span><span class="n">delete_index</span><span class="p">):</span>
        <span class="k">del</span> <span class="n">read_list</span><span class="p">[</span><span class="n">del_ind</span><span class="p">]</span> <span class="c1">#DELETES THE INDEX WHICH IS NOW ADDED UP TO THE LAST INDEX</span>
        <span class="k">del</span> <span class="n">insertion_list</span><span class="p">[</span><span class="n">del_ind</span><span class="p">]</span> <span class="c1">#DELETES THE SAME INDICES IN THE INSERTION LISTS.</span>
        <span class="k">del</span> <span class="n">number_of_duplicates_list</span><span class="p">[</span><span class="n">del_ind</span><span class="p">]</span>

    <span class="n">readspertransposon_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">/</span><span class="n">y</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">read_list</span><span class="p">,</span> <span class="n">number_of_duplicates_list</span><span class="p">)]</span> <span class="c1">#DIVIDE THE NUMBER OF READS BY THE NUMBER OF TRANSPOSONS</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">readspertransposon_list</span> <span class="o">=</span> <span class="n">read_list</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="create-list-of-all-locations-in-the-gene">
<h3>Create list of all locations in the gene<a class="headerlink" href="#create-list-of-all-locations-in-the-gene" title="Permalink to this headline">¶</a></h3>
<p>Create a list with the same length as the number of basepairs in the gene where for each location the number of reads are stored.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gene_length</span> <span class="o">=</span> <span class="n">gene_end</span><span class="o">-</span><span class="n">gene_start</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Length of region of interest is &#39;</span><span class="p">,</span><span class="n">gene_length</span><span class="p">)</span>
<span class="n">insertion_roi_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">gene_start</span><span class="p">,</span><span class="n">gene_end</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<span class="n">reads_roi_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">gene_length</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>

<span class="n">read_index</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">position</span> <span class="ow">in</span> <span class="n">insertion_list</span><span class="p">:</span>
    <span class="n">roi_index</span> <span class="o">=</span> <span class="n">insertion_roi_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">position</span><span class="p">)</span>
    <span class="n">reads_roi_list</span><span class="p">[</span><span class="n">roi_index</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">readspertransposon_list</span><span class="p">[</span><span class="n">read_index</span><span class="p">])</span>
    <span class="n">read_index</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Length of region of interest is  1655
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="statistics-and-prepare-for-plotting">
<h3>Statistics and prepare for plotting<a class="headerlink" href="#statistics-and-prepare-for-plotting" title="Permalink to this headline">¶</a></h3>
<p>Some statistical values are determined, like the average number insertions as well as the median, the percentiles and the coverage. Also the number of basepairs between the insertions are determined here.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bp_between_tn_insertions_chr_dict</span> <span class="o">=</span> <span class="n">statistics_perchromosome</span><span class="o">.</span><span class="n">chromosome_insertion_periodicity</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">,</span><span class="n">bed_file</span><span class="p">)</span>
<span class="n">insertion_chromosome_avgperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">bp_between_tn_insertions_chr_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">))</span>
<span class="n">insertion_chromosome_stdperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanstd</span><span class="p">(</span><span class="n">bp_between_tn_insertions_chr_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">))</span>
<span class="n">insertion_chromosome_firstquartileperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanpercentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions_chr_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">),</span><span class="mi">25</span><span class="p">)</span>
<span class="n">insertion_chromosome_medperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanpercentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions_chr_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">),</span><span class="mi">50</span><span class="p">)</span>
<span class="n">insertion_chromosome_thirdquartileperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanpercentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions_chr_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">),</span><span class="mi">75</span><span class="p">)</span>

<span class="n">coverage_percentage</span> <span class="o">=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">read_list</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">insertion_roi_list</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span>

<span class="k">if</span> <span class="n">insertion_list</span> <span class="o">!=</span> <span class="p">[]:</span>
    <span class="n">bp_between_tn_insertions</span> <span class="o">=</span> <span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">y</span><span class="o">-</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">insertion_list</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">insertion_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:])]</span>
    <span class="n">bp_between_tn_insertions</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">insertion_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">gene_start</span><span class="p">)</span> <span class="c1">#ADD START OF GENE (bp=0)</span>
    <span class="n">bp_between_tn_insertions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_end</span> <span class="o">-</span> <span class="n">insertion_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="c1">#ADD END OF GENE (bp=INDEX LAST TN - GENE LENGTH)</span>

    <span class="n">max_empty_region</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">)</span>
    
    <span class="n">insertion_avgperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">)</span>
    <span class="n">insertion_stdperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanstd</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">)</span>
    <span class="n">insertion_firstquartileperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanpercentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">,</span><span class="mi">25</span><span class="p">)</span>
    <span class="n">insertion_medperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanpercentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">,</span><span class="mi">50</span><span class="p">)</span>
    <span class="n">insertion_thirdquartileperiodicity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanpercentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">,</span><span class="mi">75</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">max_empty_region</span> <span class="o">=</span> <span class="n">gene_length</span>
    <span class="n">bp_between_tn_insertions</span> <span class="o">=</span> <span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">y</span><span class="o">-</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">insertion_list</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">insertion_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:])]</span>
    <span class="n">insertion_avgperiodicity</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">insertion_medperiodicity</span> <span class="o">=</span> <span class="mi">0</span>

<span class="k">if</span> <span class="n">insertion_list</span> <span class="o">!=</span> <span class="p">[]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Percentage of coverage is </span><span class="si">%.2f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">coverage_percentage</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Mean transposon insertion frequency in gene is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_avgperiodicity</span><span class="p">,</span> <span class="n">insertion_stdperiodicity</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Mean transposon insertion frequency in chromosome is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_chromosome_avgperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_stdperiodicity</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Quartiles transposon insertion frequency in gene is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_firstquartileperiodicity</span><span class="p">,</span> <span class="n">insertion_medperiodicity</span><span class="p">,</span> <span class="n">insertion_thirdquartileperiodicity</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Quartiles transposon insertion frequency in chromosome is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_chromosome_firstquartileperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_medperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_thirdquartileperiodicity</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>
Percentage of coverage is 1.09

Mean transposon insertion frequency in gene is 87.11, 101.69 
Mean transposon insertion frequency in chromosome is 22.65, 51.73

Quartiles transposon insertion frequency in gene is 7.50, 47.00, 128.00 
Quartiles transposon insertion frequency in chromosome is 3.00, 7.00, 20.00 
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="binning">
<h3>Binning<a class="headerlink" href="#binning" title="Permalink to this headline">¶</a></h3>
<p>The gene is divided in several bins. The width of the bins are determined to include 8 transposons. So if there are many transposons close together, the bin width will be small whereas insertions that are well separated results in large bins.
To account for the situation that are very little insertions or no insertions at all, a maximum bin width is determined to be 8 times the average distance between insertions for the entire chromosome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reads_roi_binnedlist</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#STORES THE HEIGHT OF THE BARS</span>
<span class="n">insertion_roi_binnedlist</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#STORES THE X POSITION OF THE BARS</span>
<span class="n">bin_width</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#STORES THE WIDTH OF THE BARS</span>

<span class="n">currentbin_reads_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">currentbin_insertion_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">reads_currentbin</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">insertion_counter</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">ins</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">insertion_roi_list</span><span class="p">)):</span>
    <span class="n">currentbin_reads_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reads_roi_list</span><span class="p">[</span><span class="n">ins</span><span class="p">])</span>
    <span class="n">currentbin_insertion_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">insertion_roi_list</span><span class="p">[</span><span class="n">ins</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">reads_roi_list</span><span class="p">[</span><span class="n">ins</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">reads_currentbin</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">if</span> <span class="n">reads_currentbin</span> <span class="o">==</span> <span class="mi">8</span><span class="p">:</span> <span class="c1">#STOP IF THERE ARE 8 INSERTIONS FOUND</span>
        <span class="n">reads_roi_binnedlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">currentbin_reads_list</span><span class="p">))</span>
        <span class="n">bin_width</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">currentbin_insertion_list</span><span class="p">)</span> <span class="o">-</span> <span class="nb">min</span><span class="p">(</span><span class="n">currentbin_insertion_list</span><span class="p">))</span>
        <span class="n">insertion_roi_binnedlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">insertion_roi_list</span><span class="p">[</span><span class="n">ins</span><span class="p">]</span><span class="o">-</span><span class="n">bin_width</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
        
        <span class="n">currentbin_reads_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">currentbin_insertion_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">reads_currentbin</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">insertion_counter</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">elif</span> <span class="n">insertion_counter</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">insertion_chromosome_avgperiodicity</span><span class="o">*</span><span class="mi">8</span><span class="p">):</span> <span class="c1">#STOP IF THE LENGTH OF THE CURRENTBIN EXCEEDS AVERAGE PERIODICITY OF THE CHROMOSOME * 8</span>
        <span class="n">reads_roi_binnedlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">currentbin_reads_list</span><span class="p">))</span>
        <span class="n">bin_width</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">currentbin_insertion_list</span><span class="p">)</span> <span class="o">-</span> <span class="nb">min</span><span class="p">(</span><span class="n">currentbin_insertion_list</span><span class="p">))</span>
        <span class="n">insertion_roi_binnedlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">insertion_roi_list</span><span class="p">[</span><span class="n">ins</span><span class="p">]</span><span class="o">-</span><span class="n">bin_width</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
        
        <span class="n">currentbin_reads_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">currentbin_insertion_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">reads_currentbin</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">insertion_counter</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">elif</span> <span class="n">gene_start</span><span class="o">+</span><span class="n">ins</span> <span class="o">==</span> <span class="n">gene_end</span><span class="p">:</span> <span class="c1">#ACCOUNT FOR THE FINAL BIN IN THE GENE.</span>
        <span class="n">reads_roi_binnedlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">currentbin_reads_list</span><span class="p">))</span>
        <span class="n">bin_width</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">currentbin_insertion_list</span><span class="p">)</span> <span class="o">-</span> <span class="nb">min</span><span class="p">(</span><span class="n">currentbin_insertion_list</span><span class="p">))</span>
        <span class="n">insertion_roi_binnedlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">insertion_roi_list</span><span class="p">[</span><span class="n">ins</span><span class="p">]</span><span class="o">-</span><span class="n">bin_width</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">insertion_counter</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="plotting">
<h3>Plotting<a class="headerlink" href="#plotting" title="Permalink to this headline">¶</a></h3>
<p>Make the bar plot and the violinplot.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">gene_name</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Plotting reads for gene &#39;</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="s1">&#39;...&#39;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Plotting reads in range &#39;</span><span class="p">,</span> <span class="n">gene_start</span><span class="p">,</span> <span class="s1">&#39;..&#39;</span><span class="p">,</span> <span class="n">gene_end</span><span class="p">,</span> <span class="s1">&#39;in chromosome &#39;</span><span class="p">,</span> <span class="n">gene_chr</span><span class="p">,</span> <span class="s1">&#39;...&#39;</span><span class="p">)</span>

<span class="n">text_size</span> <span class="o">=</span> <span class="mi">12</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span><span class="c1">#(figsize=(19,9))</span>
<span class="n">grid</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">grid</span><span class="p">[</span><span class="mi">0</span><span class="p">,:</span><span class="mi">2</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">insertion_roi_binnedlist</span><span class="p">,</span><span class="n">reads_roi_binnedlist</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="n">bin_width</span><span class="p">,</span><span class="n">facecolor</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">126.0</span><span class="p">,</span><span class="mf">164.0</span><span class="p">,</span><span class="mf">179.0</span><span class="p">])</span><span class="o">/</span><span class="mi">255</span><span class="p">,</span><span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_axisbelow</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="k">if</span> <span class="n">gene_name</span> <span class="o">!=</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">text_size</span><span class="p">)</span>
<span class="k">elif</span> <span class="n">region</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;IV&#39;</span><span class="p">,</span><span class="mi">46271</span><span class="p">,</span><span class="mi">48031</span><span class="p">]:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;HO-locus&#39;</span><span class="p">,</span> <span class="n">fontweight</span><span class="o">=</span><span class="s1">&#39;bold&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">text_size</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">)</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_start</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_end</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Basepair position in chromosome &#39;</span><span class="o">+</span> <span class="n">gene_chr</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">text_size</span><span class="p">)</span> 
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Read/Tn&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">text_size</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">gene_start</span><span class="p">,</span><span class="n">gene_end</span><span class="p">)</span>

<span class="k">if</span> <span class="n">gene_name</span> <span class="o">!=</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">gene_name</span> <span class="o">!=</span> <span class="s1">&#39;HOlocus&#39;</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">insertion_list</span> <span class="o">!=</span> <span class="p">[]:</span>
        <span class="n">textstr</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="sa">r</span><span class="s1">&#39;Reading orientation of gene: &#39;</span> <span class="o">+</span> <span class="n">gene_orien</span><span class="p">,</span>
                            <span class="sa">r</span><span class="s1">&#39;Transposon coverage = </span><span class="si">%.2f</span><span class="s1"> percent&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">coverage_percentage</span><span class="p">),</span>
                            <span class="sa">r</span><span class="s1">&#39;Mean transposon insertion frequency in gene is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_avgperiodicity</span><span class="p">,</span> <span class="n">insertion_stdperiodicity</span><span class="p">),</span>
                            <span class="sa">r</span><span class="s1">&#39;Mean transposon insertion frequency in chromosome is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_chromosome_avgperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_stdperiodicity</span><span class="p">),</span>
                            <span class="sa">r</span><span class="s1">&#39;Quartiles transposon insertion frequency in gene is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_firstquartileperiodicity</span><span class="p">,</span> <span class="n">insertion_medperiodicity</span><span class="p">,</span> <span class="n">insertion_thirdquartileperiodicity</span><span class="p">),</span>
                            <span class="sa">r</span><span class="s1">&#39;Quartiles transposon insertion frequency in chromosome is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_chromosome_firstquartileperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_medperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_thirdquartileperiodicity</span><span class="p">)</span>
                            <span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">textstr</span> <span class="o">=</span> <span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Reading orientation of gene: &#39;</span> <span class="o">+</span> <span class="n">gene_orien</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">textstr</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">((</span><span class="sa">r</span><span class="s1">&#39;Transposon coverage = </span><span class="si">%.2f</span><span class="s1"> percent&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">coverage_percentage</span><span class="p">),</span>
                        <span class="sa">r</span><span class="s1">&#39;Mean transposon insertion frequency in gene is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_avgperiodicity</span><span class="p">,</span> <span class="n">insertion_stdperiodicity</span><span class="p">),</span>
                        <span class="sa">r</span><span class="s1">&#39;Mean transposon insertion frequency in chromosome is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_chromosome_avgperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_stdperiodicity</span><span class="p">),</span>
                        <span class="sa">r</span><span class="s1">&#39;Quartiles transposon insertion frequency in gene is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_firstquartileperiodicity</span><span class="p">,</span> <span class="n">insertion_medperiodicity</span><span class="p">,</span> <span class="n">insertion_thirdquartileperiodicity</span><span class="p">),</span>
                        <span class="sa">r</span><span class="s1">&#39;Quartiles transposon insertion frequency in chromosome is </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1">, </span><span class="si">%.2f</span><span class="s1"> &#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">insertion_chromosome_firstquartileperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_medperiodicity</span><span class="p">,</span> <span class="n">insertion_chromosome_thirdquartileperiodicity</span><span class="p">)</span>
                        <span class="p">))</span>
<span class="n">props</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">boxstyle</span><span class="o">=</span><span class="s1">&#39;round&#39;</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.05</span><span class="p">,</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">textstr</span><span class="p">,</span> <span class="n">transform</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">transAxes</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">text_size</span><span class="p">,</span>
    <span class="n">verticalalignment</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span><span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="n">props</span><span class="p">)</span>

<span class="k">for</span> <span class="n">ins</span> <span class="ow">in</span> <span class="n">insertion_list</span><span class="p">:</span> <span class="c1">#PLOT TICKS ON THE X AXIS TO INDICATE THE EXACT INSERTION LOCATIONS</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">ins</span><span class="p">,</span> <span class="n">ymax</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>



<span class="c1"># COMPARE DISTRIBUTION OF BASEPAIRS BETWEEN INSERTIONS FOR THE CHROMOSOME AND THE GENE USING VIOLINPLOT</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">grid</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">])</span>

<span class="k">if</span> <span class="n">gene_name</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">region</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;IV&#39;</span><span class="p">,</span><span class="mi">46271</span><span class="p">,</span><span class="mi">48031</span><span class="p">]:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="s1">&#39;HOlocus&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">)</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_start</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_end</span><span class="p">)</span>

<span class="n">bp_between_tn_insertions_dict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">bp_between_tn_insertions_dict</span><span class="p">[</span><span class="n">gene_chr</span><span class="p">]</span> <span class="o">=</span> <span class="n">bp_between_tn_insertions</span>

<span class="n">df_chr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">bp_between_tn_insertions_chr_dict</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">bp_between_tn_insertions_dict</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
    <span class="n">df_concat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span><span class="n">df_chr</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">names_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">bp_between_tn_insertions_chr_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">gene_chr</span><span class="p">))</span>
    <span class="n">df_concat</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">names_list</span>
    <span class="n">xlabel</span><span class="o">=</span><span class="n">gene_name</span><span class="o">+</span><span class="s1">&#39; | &#39;</span><span class="o">+</span><span class="n">gene_chr</span>
    <span class="n">df_concat</span><span class="p">[</span><span class="n">xlabel</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">df_concat</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;bp between tn&#39;</span><span class="p">,</span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="n">xlabel</span><span class="p">]</span>

    <span class="n">sb</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s2">&quot;pastel&quot;</span><span class="p">,</span> <span class="n">color_codes</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">sb</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df_concat</span><span class="p">,</span><span class="n">x</span><span class="o">=</span><span class="n">xlabel</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;bp between tn&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;label&#39;</span><span class="p">,</span> <span class="n">inner</span><span class="o">=</span><span class="s1">&#39;quartile&#39;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="s1">&#39;width&#39;</span><span class="p">,</span> <span class="n">gridsize</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">split</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cut</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">:</span> <span class="s2">&quot;r&quot;</span><span class="p">})</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">max_empty_region</span><span class="o">+</span><span class="mi">10</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">sb</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df_chr</span><span class="p">,</span> <span class="n">inner</span><span class="o">=</span><span class="s1">&#39;quartile&#39;</span><span class="p">,</span> <span class="n">orien</span><span class="o">=</span><span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="s1">&#39;width&#39;</span><span class="p">,</span> <span class="n">gridsize</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">cut</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="k">if</span> <span class="n">savefigure</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">save_figure_path</span><span class="p">,</span> <span class="n">gene_name</span><span class="o">+</span><span class="s1">&#39;_TnInsertions.png&#39;</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Figure saved at &#39;</span><span class="p">,</span> <span class="n">save_figure_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Plotting reads for gene  BEM1 ...
</pre></div>
</div>
<img alt="_images/gene_reads_20_1.png" src="_images/gene_reads_20_1.png" />
</div>
</div>
</div>
</div>
<div class="section" id="interpretation">
<h2>Interpretation<a class="headerlink" href="#interpretation" title="Permalink to this headline">¶</a></h2>
<p>The resulting barplot shows the insertions (represented by the small black bars) along the gene. The blue bars indicate the number of reads per transposon in that bin. This shows how the insertions and reads are distributed along the gene.
The reading orientation is indicated by a + (forward reading) or - (reverse reading). The numbers for the mean are given together with the standard deviation (note that this is typically very large). Therefore a more reliable measure might be the percentiles given as the 25th, 50th and the 75th percentile respectively.</p>
<p>The violinplot shows the distribution of the amount of basepairs between subsequent insertions for both the gene (green) and the chromomsome the gene is in (red). A significant difference in the median basepairs between insertions might indicate potential essential genes.</p>
</div>
<div class="section" id="bibliography">
<h2>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Michel, A. H., Hatakeyama, R., Kimmig, P., Arter, M., Peter, M., Matos, J., … &amp; Kornmann, B. (2017). Functional mapping of yeast genomes by saturated transposition. Elife, 6, e23570.</p></li>
</ul>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "leilaicruz/SATAY-jupyter-book",
            ref: "gh-pages",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "."
        }
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="MysT_Notebooks.html" title="previous page">Notebooks written entirely in markdown</a>
    <a class='right-next' id="next-link" href="statistics_pergene.html" title="next page">Statistics per gene</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Gregory van Beek<br/>
        
            &copy; Copyright GPL License 3.0.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    <script src="_static/js/index.js"></script>
    
  </body>
</html>