
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Status of the project &#8212; SATAY pipeline at Delft :)</title>
    
  <link href="_static/css/theme.css" rel="stylesheet" />
  <link href="_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="_static/genetica_modified_neon.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Data analysis steps" href="00-Intro/data-analysis-workflow.html" />
    <link rel="prev" title="Documentation for processing SATAY" href="documentation_satay.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      <img src="_static/genetica_modified_neon.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">SATAY pipeline at Delft :)</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="documentation_satay.html">
   Documentation for processing SATAY
   <!-- omit in toc -->
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Status of the project
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="00-Intro/data-analysis-workflow.html">
   Data analysis steps
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="00-Intro/flowchart.html">
     Flowchart of the pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="00-Intro/Installation_Guide_SATAY_Analysis_Software.html">
     Software installation guide for whole genome sequencing and transposon mapping
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="00-Intro/satay_analysis_notes.html">
     All what you need to know from SATAY
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="01-bash-GUI/processing_workflow.html">
   Code for the bash GUI
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="02-postprocessing-scripts/00-Introduction.html">
   Python scripts
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="02-postprocessing-scripts/01-clean_bedwigfiles.html">
     Cleaning bed and wig files
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="02-postprocessing-scripts/02-transposonread_profileplot_genome.html">
     Transposon profile for the whole genome
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="02-postprocessing-scripts/03-transposonread_profileplot.html">
     Transposon profile for a chromosome of interest
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="02-postprocessing-scripts/04-scatterplot_genes.html">
     Create a scatterplot for all genes and all essential genes.
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="02-postprocessing-scripts/05-volcanoplot.html">
     Volcano plots to two different libraries
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="02-postprocessing-scripts/06-genomicfeatures_dataframe.html">
     Searching for genomic features  of transposition events
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="03-docker-doc/00-Docker-Users.html">
   Build your own docker container with ours satay pipeline
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/Introduction.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/SATAY-LL/SATAY-jupyter-book"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/SATAY-LL/SATAY-jupyter-book/issues/new?title=Issue%20on%20page%20%2FIntroduction.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Status of the project
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#next-steps">
     Next steps
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#what-do-we-have-currently">
   What do we have currently?
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#software-license">
   Software License
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#current-contact">
   Current Contact
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#acknowledgements">
   Acknowledgements üíô
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <p>Hello there üòÑ Welcome to this Jupyter Book instance to document our data processing pipeline to analyze data from the technique : <strong>Sa</strong>turated <strong>T</strong>ransposition <strong>A</strong>nalysis in <strong>Y</strong>east.</p>
<div class="section" id="status-of-the-project">
<h1>Status of the project<a class="headerlink" href="#status-of-the-project" title="Permalink to this headline">¬∂</a></h1>
<ul class="simple">
<li><p>Beta release on Zenodo</p></li>
<li><p>Lacking automated tests</p></li>
</ul>
<div class="section" id="next-steps">
<h2>Next steps<a class="headerlink" href="#next-steps" title="Permalink to this headline">¬∂</a></h2>
<!-- 
1. Make the software package ready to be published in [Journal of Open Source Software](https://joss.readthedocs.io/en/latest/index.html)

    - Automated testing with continuous integration
    - Comand line interface with a pip install mode
    - GitFlow implementation for contributors to the software -->
<!-- 3.  Make a plan how to define trimming and alignment settings.
Currently the trimming and processing options are chosen by trial-and-error.
Each dataset requires unique processing and therefore no fixed settings for the trimming and alignment can be determined, but potentially some guidelines can be created that helps for the processing. -->
<!-- 4. The code is relying on software third party software tools (e.g. for the trimming, alignment and quality checking).
This currently requires to be installed manually and in the [workflow](https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/blob/master/satay.sh) paths need to be set to some these tools (in satay.sh, the paths are all defined in the beginning of the script in the `DEFINE PATHS` section). This is not convenient for other users and is prone to errors.
Therefore a package need to be created that installs the third party software tools at a fixed location relative to the workflow.
Then the paths doesn't have to set manually by the user and makes it easier to set up and use. -->
<!-- 
5. Implementing unit testing in the Github repository. This allows for automatic testing of the codes after changes have been made in the repository and should ensure the outcome is still correct.
Most importantly this needs to be implemented in the [transposonmapping_satay.py](https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/blob/master/python_transposonmapping/transposonmapping_satay.py) code (the final step of the processing pipeline) and the depending [python modules](https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/tree/master/python_transposonmapping/python_modules).
The other processing steps are all relying on third party software that probably already some testing implemented. -->
<!-- 6. For the transposon mapping in the pipeline [a custom python script](https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/blob/master/python_transposonmapping/transposonmapping_satay.py) is created that inputs a bam file and outputs lists of insertion locations and the corresponding number of reads. This script is based on the [matlab code](https://sites.google.com/site/satayusers/complete-protocol/bioinformatics-analysis/matlab-script) from the Kornmann lab.
The matlab code contained some bugs that were solved in the python script (see the [satay forum](https://groups.google.com/g/satayusers/search?q=matlab)). Therefore some small differences may be present between the matlab processing and the python processing. But recently, some more differences were found that could not be explained by the matlab bugs.
The exact cause is yet unclear, but has maybe to do with reading the samflag in the python script (line 183 in the original [transposonmapping_satay.py]((https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/blob/master/python_transposonmapping/transposonmapping_satay.py)) code which uses the [samflag.py](https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/blob/master/python_transposonmapping/python_modules/samflag.py) module).
The insertions found by the transposon mapping python scripts can be compared with IGV that loads the corresponding bam file.
This shows the read locations so the output of the transposon mappping script can be manually checked. -->
<!-- 7.  Currently [BWA](http://bio-bwa.sourceforge.net/bwa.shtml) is used for the alignment of the reads.
This tools work fine, but the Kornmann lab (and many other labs) use [Bowtie](http://bowtie-bio.sourceforge.net/index.shtml).
This is already installed on the Linux machine, but needs to be implemented in the [workflow](https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/blob/master/satay.sh).
The current aligner is implemented in the section `SEQUENCE ALIGNMENT`.
Note to set the option for paired-end and single-end alignment for bowtie (defined in the variable `${paired}`). -->
<!-- 8. Allow mulitple files to be processed at once using a for-loop in the [bash script](https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/blob/master/satay.sh).
Currently only one dataset at the time can be processed, but this can be extended to allow a series of datasets to be processed in a for-loop.
To allow for this, the bash script satay.sh should be altered where the code between line 285 and line 680 (of the original code) should be placed in a loop over all file paths.
Ideally each dataset is located in a separate folder where the output folders and files are stored for that particular dataset.
If multiple datasets are located in the same folder, it might happen that things are being overwritten.
It is possible to put all datasets to be processed in the same folder, but then in the code all these datasets should be automatically be moved to individual folders.
Note to pay attention that the paths to the datasets are all correct in the for-loop. -->
<!-- 9. Automated documentation based on the help texts from the python scripts. -->
<div class="sphinx-bs container pb-4 full-width text-center docutils">
<div class="row docutils">
<div class="d-flex col-lg-6 px-2 py-2 docutils">
<div class="card w-100 docutils">
<div class="card-header docutils">
<p class="card-text"><strong>Publication in JOSS</strong> ‚úç</p>
</div>
<div class="card-body docutils">
<p class="card-text">Make the software package ready to be published in <a class="reference external" href="https://joss.readthedocs.io/en/latest/index.html">Journal of Open Source Software</a></p>
  <!-- - Automated testing with continuous integration
    - Comand line interface with a pip install mode
    - GitFlow implementation for contributors to the software -->
</div>
</div>
</div>
<div class="d-flex col-lg-6 px-2 py-2 docutils">
<div class="card w-100 docutils">
<div class="card-header docutils">
<p class="card-text"><strong>Data sequencing processing guidelines</strong> üöÄ</p>
</div>
<div class="card-body docutils">
<p class="card-text">Each dataset requires unique processing and therefore no fixed settings for the trimming and alignment can be determined, but potentially some guidelines can be created that helps for the processing.</p>
</div>
</div>
</div>
<div class="d-flex col-lg-6 px-2 py-2 docutils">
<div class="card w-100 docutils">
<div class="card-header docutils">
<p class="card-text"><strong>Unit testing continous integration</strong> ‚ú®</p>
</div>
<div class="card-body docutils">
<p class="card-text">Unit testing implemented in the main functions to make the pipeline more maintanable and to prevent crashes when customizing it.</p>
</div>
</div>
</div>
<div class="d-flex col-lg-6 px-2 py-2 docutils">
<div class="card w-100 docutils">
<div class="card-header docutils">
<p class="card-text"><strong>Multiple files analysis</strong> üéÅ</p>
</div>
<div class="card-body docutils">
<p class="card-text">Allow mulitple files to be processed at once using a for-loop in the bash script.
Currently only one dataset at the time can be processed, but this can be extended to allow a series of datasets to be processed in a for-loop.</p>
<!-- - To allow for this, the bash script satay.sh should be altered where the code between line 285 and line 680 (of the original code) should be placed in a loop over all file paths.
Ideally each dataset is located in a separate folder where the output folders and files are stored for that particular dataset.
If multiple datasets are located in the same folder, it might happen that things are being overwritten.
It is possible to put all datasets to be processed in the same folder, but then in the code all these datasets should be automatically be moved to individual folders.
Note to pay attention that the paths to the datasets are all correct in the for-loop. -->
</div>
</div>
</div>
<div class="d-flex col-lg-6 px-2 py-2 docutils">
<div class="card w-100 docutils">
<div class="card-header docutils">
<p class="card-text"><strong>Python package</strong> üõ∞</p>
</div>
<div class="card-body docutils">
<p class="card-text">Build a python package that is pip installable for the ease of use for other researchers.</p>
</div>
</div>
</div>
<div class="d-flex col-lg-6 px-2 py-2 docutils">
<div class="card w-100 docutils">
<div class="card-header docutils">
<p class="card-text"><strong>Tutorials</strong> üìù</p>
</div>
<div class="card-body docutils">
<p class="card-text">Make tutorials of usage of the pipeline</p>
</div>
</div>
</div>
<div class="d-flex col-lg-6 px-2 py-2 docutils">
<div class="card w-100 docutils">
<div class="card-body docutils">
</div>
</div>
</div>
</div>
</div>
<div class="tip full-width admonition">
<p class="admonition-title">Learn more and get involved</p>
<dl class="simple myst">
<dt>üí° <a class="reference external" href="https://github.com/SATAY-LL/LaanLab-SATAY-DataAnalysis/issues">Open an issue</a></dt><dd><p>We track enhancement requests, bug-reports, and to-do items via GitHub issues.</p>
</dd>
<dt>üôå <a class="reference external" href="https://github.com/SATAY-LL/LaanLab-SATAY-DataAnalysis/pulls">Send a pull request</a></dt><dd><p>We appreciate that contributors send a pull request to improve the code base or introduce a new feature.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="section" id="what-do-we-have-currently">
<h1>What do we have currently?<a class="headerlink" href="#what-do-we-have-currently" title="Permalink to this headline">¬∂</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://zenodo.org/badge/latestdoi/248577762"><img alt="DOI" src="https://zenodo.org/badge/248577762.svg" /></a></p></li>
<li><p><a class="reference external" href="https://github.com/leilaicruz/LaanLab-SATAY-DataAnalysis/archive/refs/tags/v1.0-beta.1.zip">Download here the Public release v1.0.beta-1</a></p></li>
</ul>
</div>
<div class="section" id="software-license">
<h1>Software License<a class="headerlink" href="#software-license" title="Permalink to this headline">¬∂</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://opensource.org/licenses/Apache-2.0"><img alt="License" src="https://img.shields.io/badge/License-Apache%202.0-blue.svg" /></a></p></li>
<li><p>This work is licensed under Apache 2.0 .
The 2.0 version of the Apache License, approved by the ASF in 2004, helps us achieve our goal of providing reliable and long-lived software products through collaborative open source software development.</p></li>
</ul>
</div>
<div class="section" id="current-contact">
<h1>Current Contact<a class="headerlink" href="#current-contact" title="Permalink to this headline">¬∂</a></h1>
<ul class="simple">
<li><p>Leila Inigo de la Cruz</p>
<ul>
<li><p>email: <a class="reference external" href="mailto:L&#46;M&#46;InigoDeLaCruz&#37;&#52;&#48;tudelft&#46;nl">L<span>&#46;</span>M<span>&#46;</span>InigoDeLaCruz<span>&#64;</span>tudelft<span>&#46;</span>nl</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="acknowledgements">
<h1>Acknowledgements üíô<a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¬∂</a></h1>
<ul class="simple">
<li><p>Many thanks to the group of <a class="reference external" href="https://www.bioch.ox.ac.uk/research/kornmann">Prof. Benoit Kornmann</a> , in particular to <a class="reference external" href="https://www.kornmann.group/people/agn%C3%A8s">Prof. Agnes Michel</a> , for all the nice discussions and colaboration with our team. Also thanks to them due to the <a class="reference external" href="https://groups.google.com/forum/#%21forum/satayusers">SATAY forum</a>  they keep on running for all the community of satayers üéâ .</p></li>
<li><p>Many thanks to the <a class="reference external" href="https://tu-delft-dcc.github.io/01_Welcome.html">DCC department at TU Delft</a>, specially to <a class="reference external" href="https://github.com/mwakok">Maurits Kok</a> who in his role of Research Software Engineer has supported us in developing the software package to be distributed to the scientific community.</p></li>
</ul>
<!-- :::{image} https://pbs.twimg.com/profile_images/1226944724365447169/MzFpwY5P_400x400.png
:class: float-left mr-2 rounded
:width: 100px
::: --></div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="documentation_satay.html" title="previous page">Documentation for processing SATAY <!-- omit in toc --></a>
    <a class='right-next' id="next-link" href="00-Intro/data-analysis-workflow.html" title="next page">Data analysis steps</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Gregory van Beek , Maurits Kok, Leila I√±igo de la Cruz<br/>
        
            &copy; Copyright Apache 2.0.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>