
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>2. Statistics per gene &#8212; SATAY DATA ANALYSIS LAAN-LAB TU DELFT</title>
    
  <link rel="stylesheet" href="_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx_tabs/semantic-ui-2.4.1/segment.min.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx_tabs/semantic-ui-2.4.1/menu.min.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx_tabs/semantic-ui-2.4.1/tab.min.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx_tabs/tabs.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"TeX": {"Macros": {"N": "\\mathbb{N}", "floor": ["\\lfloor#1\\rfloor", 1], "bmat": ["\\left[\\begin{array}"], "emat": ["\\end{array}\\right]"]}}, "tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="_static/Greg.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="3. Comparison of the statistics per chromosome" href="statistics_perchromosome_Compare.html" />
    <link rel="prev" title="1. Gene reads script" href="gene_reads.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="index.html">
  
  <img src="_static/Greg.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">SATAY DATA ANALYSIS LAAN-LAB TU DELFT</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="Introduction.html">
   To reproduce an example of  Jupyter book, and you are a Windows user:
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Get started
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="Intro/flowchart.html">
   1. Flowchart of the pipeline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Intro/satay_analysis_notes.html">
   2. All what you need to know from SATAY
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Intro/Installation_Guide_SATAY_Analysis_Software.html">
   3. ” Software installation guide for whole genome sequencing and transposon mapping”
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Data Stats
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="transposonmapping_satay.html">
   1. Python Translation of the Matlab code from Kornmann Lab
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Postprocessing scripts
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="gene_reads.html">
   1. Gene reads script
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   2. Statistics per gene
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="statistics_perchromosome_Compare.html">
   3. Comparison of the statistics per chromosome
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="from-reads-to-essential-genes.html">
   4. From reads and insertions to essentiality prediction using a Regression Model
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/statistics_pergene.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/leilaicruz/SATAY-jupyter-book"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/leilaicruz/SATAY-jupyter-book/issues/new?title=Issue%20on%20page%20%2Fstatistics_pergene.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#the-python-script">
   2.1. The python script
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     2.1.1. Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-list-of-all-essential-genes">
     2.1.2. Create list of all essential genes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-dictionary-with-length-all-genes">
     2.1.3. Create a dictionary with length all genes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-lists-with-information-to-be-plotted">
     2.1.4. Create lists with information to be plotted
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-dataframe">
     2.1.5. Create dataframe
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#print-statistics">
     2.1.6. Print statistics
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-histogram">
     2.1.7. Create a histogram
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-a-violin-plot">
     2.1.8. Create a violin plot
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#interpretation">
   2.2. Interpretation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bibliography">
   2.3. Bibliography
  </a>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="statistics-per-gene">
<h1><span class="section-number">2. </span>Statistics per gene<a class="headerlink" href="#statistics-per-gene" title="Permalink to this headline">¶</a></h1>
<div class="section" id="the-python-script">
<h2><span class="section-number">2.1. </span>The python script<a class="headerlink" href="#the-python-script" title="Permalink to this headline">¶</a></h2>
<p>This code determines statistics based on the results obtained in the _pergene.txt file by the Matlab code provided by the Kornmann lab.
The _pergene.txt file consists of three columns.
The rows in the first column contains all genes of * S.Cerevisiae*, the corresponding transposon count is stored in the second column and the third column contains the reads.</p>
<p>The genes in this file are split into two groups based on their essentiality.
This is determined using two files that contains all known essential genes.</p>
<p>The data is represented as a violinplot where both the essential and non-essential genes are shown.</p>
<p>Some basic statistics are determined here, but this can be extended with more if needed.</p>
<div class="section" id="input">
<h3><span class="section-number">2.1.1. </span>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>Data that needs to be loaded:</p>
<ol class="simple">
<li><p>filepath: Path to the output text file the Matlab code of Benoit that includes the read and transposon count per gene (with extension _pergene.txt)</p></li>
<li><p>filename: Name of this file.</p></li>
<li><p>normalize: Whether or not to normalize the transposon and read counts to the gene lengths (True or False)</p></li>
<li><p>essential_genes_files: List of essential genes. Can be input as multiple files which will be automatically merged.</p></li>
<li><p>gene_information_file: List of genes with their possible aliasses and lengths in amino acids.</p></li>
</ol>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>import os
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
</pre></div>
</div>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>
file_dirname = os.path.dirname(&#39;__file__&#39;)
# sys.path.insert(1,os.path.join(file_dirname,&#39;python_modules&#39;))
# os.chdir(&#39;mini_book/docs/Python_scripts/python_modules/&#39;)

filepath = os.path.join(file_dirname,&#39;satay_analysis_testdata&#39;,&#39;Output_Processing&#39;)
filename = &quot;Cerevisiae_WT2_Michel2017_trimmed1.bam_pergene.txt&quot; # CHANGE THIS TO ANY FILE YOU WANT TO ANALYSE.
normalize = &#39;True&#39;

essential_genes_files = [os.path.join(file_dirname,&#39;Python_scripts&#39;,&#39;Data_Files&#39;,&#39;Cervisiae_EssentialGenes_List_1.txt&#39;),
                        os.path.join(file_dirname,&#39;Python_scripts&#39;,&#39;Data_Files&#39;,&#39;Cervisiae_EssentialGenes_List_2.txt&#39;)]

gene_information_file = os.path.join(file_dirname,&#39;Python_scripts&#39;,&#39;Data_Files&#39;,&#39;Yeast_Protein_Names.txt&#39;)
# os.chdir(&#39;mini_book/docs/Python_scripts/&#39;)
</pre></div>
</div>
</div>
<div class="section" id="create-list-of-all-essential-genes">
<h3><span class="section-number">2.1.2. </span>Create list of all essential genes<a class="headerlink" href="#create-list-of-all-essential-genes" title="Permalink to this headline">¶</a></h3>
<p>This can be input from multiple files given in the <code class="docutils literal notranslate"><span class="pre">essential_genes_files</span></code> variable that are merged in one list.</p>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>known_essential_gene_list = []

for files in essential_genes_files:
    with open(files) as f:
        for header_lines in range(3):
            next(f)
        for lines in f:
            known_essential_gene_list.append(lines.rstrip(&#39;\n&#39;))
</pre></div>
</div>
</div>
<div class="section" id="create-a-dictionary-with-length-all-genes">
<h3><span class="section-number">2.1.3. </span>Create a dictionary with length all genes<a class="headerlink" href="#create-a-dictionary-with-length-all-genes" title="Permalink to this headline">¶</a></h3>
<p>Dictionary consists of keys which are all genes (including all their potential aliasses and different naming conventions) and of values which are the length in terms of base pairs.
The number of base pairs is calculated by multiplying the number of amino acids by 3.</p>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span># CREATE A DICTIONARY WITH ALL GENES (BOTH THE COMMON NAME AND THEIR SYSTEMATIC NAME) AND SAVE THEM WITH THEIR RESPECTIVE LENGHTS (IN TERMS OF BP WHICH IS DEFINED AS bp=aa*3)
gene_length_dict = {}
with open(gene_information_file) as f:
    lines = f.readlines()

# THE GENES START AT LINE 58 AND STOP 6 LINES BEFORE THE END OF THE FILE.
    for i in range(58,len(lines)-6):
        n=0
        l = lines[i]

# COUNT HOW MANY TIMES &#39;;&#39; OCCURS IN A LINE.
        extra_columns = l.count(&#39;;&#39;)
        l_short = &#39; &#39;.join(l.split())
        l_list = l_short.split(&#39; &#39;)

# IGNORE SEQUENCES THAT ARE NOT NEEDED.
        if l_list[1+extra_columns] == &#39;GAG&#39; or l_list[1+extra_columns] == &#39;POL&#39;:
            extra_columns = extra_columns + 1

# DETERMINE GENE LENGTH AND STORE WITH THE CORRESPONDING GENE NAME (e.g. Cdc42)
        gene_length_dict[l_list[0].strip(&#39;;&#39;)] = int(l_list[5+extra_columns])*3

# DETERMINE GENE LENGTH AND STORE WITH THE CORRESPONDING GENE SYSTEMATIC NAME (e.g. YLR229C)
        gene_length_dict[l_list[1+extra_columns]] = int(l_list[5+extra_columns])*3

# DETERMINE GENE LENGTH AND STORE WITH THE CORRESPONDING GENE ALIASES (IF PRESENT).
        if extra_columns &gt; 0:
            for n in range(extra_columns+1):
                gene_length_dict[l_list[0+n].strip(&#39;;&#39;)] = int(l_list[5+extra_columns])*3
</pre></div>
</div>
</div>
<div class="section" id="create-lists-with-information-to-be-plotted">
<h3><span class="section-number">2.1.4. </span>Create lists with information to be plotted<a class="headerlink" href="#create-lists-with-information-to-be-plotted" title="Permalink to this headline">¶</a></h3>
<p>Lists are made for</p>
<ol class="simple">
<li><p>Gene names</p></li>
<li><p>Reads per gene</p></li>
<li><p>Transposons per gene</p></li>
<li><p>Whether or not a gene is essential or not (contains only True and False statements)</p></li>
<li><p>Reads per gene normalized to the length of each gene in terms of base pairs</p></li>
<li><p>Transposons per gene normalized to the length of each gene in terms of base pairs</p></li>
</ol>
<p>The gene_counter counts the number of genes included in the analysis. The fail_counter are the number of genes that are present in the data, but could be found in the list of gene lengths and could therefore not be normalized. These genes are ignored. This should not be much more than 0.5% of the total amount of genes.</p>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>file = os.path.join(filepath,filename)

with open(file) as f:
    lines = f.readlines()
</pre></div>
</div>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>:tags: [outputPrepend]

gene_name_list = [] # LIST OF ALL GENES ANALYZED AND OUTPUT BY THE MATLAB CODE OF BENOIT.
reads_pergene_list = [] # NUMBER OF READS PER GENE AS DETERMINED BY THE MATLAB CODE OF BENOIT.
tn_pergene_list = [] # NUMBER OF TRANSPOSONS PER GENE AS DETERMINED BY THE MATLAB CODE OF BENOIT.
gene_essential_boolean_list = [] # WHICH OF THE GENES ANALYZED ARE ESSENTIAL (MARKED AS TRUE) AND NON ESSENTIAL (MARKED AS FALSE)
reads_pergene_norm_list = [] # NUMBER OF READS PER GENE AS DETERMINED BY THE MATLAB CODE OF BENOIT NORMALIZED FOR THE LENGTH OF THE GENES IN TERMS OF BASEPAIRS.
tn_pergene_norm_list = [] # NUMBER OF TRANSPOSONS PER GENE AS DETERMINED BY THE MATLAB CODE OF BENOIT NORMALIZED FOR THE LENGTH OF THE GENES IN TERMS OF BASEPAIRS.

gene_counter = 0
fail_counter = 0
failed_genes_list = []
for line in lines[1:]:
    split_line = line.split(&#39;\t&#39;)
    gene_name = split_line[0].rstrip()


    if len(split_line) &gt; 1:
        tn_pergene = int(split_line[1])
    else:
        tn_pergene = np.nan


    if len(split_line) &gt; 2:
        reads_pergene = int(split_line[2])
    else:
        reads_pergene = np.nan

    reads_pergene_list.append(reads_pergene)
    tn_pergene_list.append(tn_pergene)


#!!! GENES WITH ALIASES ARE COUNTED DOUBLE IF ONE OR MORE ALIASES ARE PRESENT IN THE ESSENTIAL GENES LIST !!!
    if normalize == False:
        if gene_name in known_essential_gene_list:
            gene_essential_boolean_list.append(True) 
        else:
            gene_essential_boolean_list.append(False)
        gene_name_list.append(gene_name)

    else:
        gene_length = gene_length_dict.get(gene_name)
        try:
            reads_pergene_norm_list.append(reads_pergene/gene_length)
            tn_pergene_norm_list.append(tn_pergene/gene_length)

            if gene_name in known_essential_gene_list:
                gene_essential_boolean_list.append(True) 
            else:
                gene_essential_boolean_list.append(False)
            gene_name_list.append(gene_name)

            gene_counter = gene_counter + 1
        except:
            fail_counter = fail_counter + 1
            failed_genes_list.append(gene_name)

print(&#39;Number of genes analyzed: &#39;,gene_counter)
print(&#39;Number of genes not found: %.0f (%.2f percent) &#39;% (fail_counter, fail_counter/(gene_counter+fail_counter)*100))
print(&#39;Genes not found: &#39;,failed_genes_list)
</pre></div>
</div>
</div>
<div class="section" id="create-dataframe">
<h3><span class="section-number">2.1.5. </span>Create dataframe<a class="headerlink" href="#create-dataframe" title="Permalink to this headline">¶</a></h3>
<p>This puts the data in the form of a dataframe (part of the <code class="docutils literal notranslate"><span class="pre">pandas</span></code> package). This is necessary for plotting.</p>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>if normalize == False:
    genes = {&#39;Gene_name&#39;: gene_name_list,
            &#39;Transposons_per_gene&#39;: tn_pergene_list,
            &#39;Reads_per_gene&#39;: reads_pergene_list,
            &#39;Essential_gene&#39;: gene_essential_boolean_list
            }
    df = pd.DataFrame(genes,columns=[&#39;Gene_name&#39;,&#39;Transposons_per_gene&#39;,&#39;Reads_per_gene&#39;, &#39;Essential_gene&#39;])
else:
    genes = {&#39;Gene_name&#39;: gene_name_list,
            &#39;Transposon_density_per_gene&#39;: tn_pergene_norm_list,
            &#39;Read_density_per_gene&#39;: reads_pergene_norm_list,
            &#39;Essential_gene&#39;: gene_essential_boolean_list
            }
    df = pd.DataFrame(genes,columns=[&#39;Gene_name&#39;,&#39;Transposon_density_per_gene&#39;,&#39;Read_density_per_gene&#39;, &#39;Essential_gene&#39;])
</pre></div>
</div>
</div>
<div class="section" id="print-statistics">
<h3><span class="section-number">2.1.6. </span>Print statistics<a class="headerlink" href="#print-statistics" title="Permalink to this headline">¶</a></h3>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>print(&#39;Average number of reads all genes: %.0f &#39; % np.nanmean(reads_pergene_list))
print(&#39;Median number of reads all genes: %.0f &#39; % np.nanmedian(reads_pergene_list))
print(&#39;Standard Deviation number of reads all genes: %.0f &#39; % np.nanstd(reads_pergene_list))
print(&#39;&#39;)

print(&#39;Average number of transposons all genes: %.0f &#39; % np.nanmean(tn_pergene_list))
print(&#39;Median number of transposons all genes: %.0f &#39; % np.nanmedian(tn_pergene_list))
print(&#39;Standard Deviation number of transposons all genes: %.0f &#39; % np.nanstd(tn_pergene_list))
print(&#39;&#39;)
print(&#39;&#39;)



indices_essential_genes_list = [i for i, x in enumerate(gene_essential_boolean_list) if x == True]
reads_peressentialgene_list = [reads_pergene_list[i] for i in indices_essential_genes_list]
print(&#39;Average number of reads of essential genes: %.0f &#39; % np.nanmean(reads_peressentialgene_list))
print(&#39;Median number of reads of essential genes: %.0f &#39; % np.nanmedian(reads_peressentialgene_list))
print(&#39;Maximum number of reads of essential genes: %.0f &#39; % np.nanmax(reads_peressentialgene_list))
print(&#39;Minimum number of reads of essential genes: %.0f &#39; % np.nanmin(reads_peressentialgene_list))
print(&#39;&#39;)

indices_nonessential_genes_list = [i for i, x in enumerate(gene_essential_boolean_list) if x == False]
reads_pernonessentialgene_list = [reads_pergene_list[i] for i in indices_nonessential_genes_list]
print(&#39;Average number of reads of nonessential genes: %.0f &#39; % np.nanmean(reads_pernonessentialgene_list))
print(&#39;Median number of reads of nonessential genes: %.0f &#39; % np.nanmedian(reads_pernonessentialgene_list))
print(&#39;Maximum number of reads of nonessential genes: %.0f &#39; % np.nanmax(reads_pernonessentialgene_list))
print(&#39;Minimum number of reads of nonessential genes: %.0f &#39; % np.nanmin(reads_pernonessentialgene_list))
print(&#39;&#39;)
print(&#39;&#39;)



tn_peressentialgene_list = [tn_pergene_list[i] for i in indices_essential_genes_list]
print(&#39;Average number of transposons of essential genes: %.0f &#39; % np.nanmean(tn_peressentialgene_list))
print(&#39;Median number of transposons of essential genes: %.0f &#39; % np.nanmedian(tn_peressentialgene_list))
print(&#39;Standard deviation number of transposons of essential genes: %.0f &#39; % np.nanstd(tn_peressentialgene_list))
print(&#39;Maximum number of transposons of essential genes: %.0f &#39; % np.nanmax(tn_peressentialgene_list))
print(&#39;Minimum number of transposons of essential genes: %.0f &#39; % np.nanmin(tn_peressentialgene_list))
print(&#39;&#39;)

tn_pernonessentialgene_list = [tn_pergene_list[i] for i in indices_nonessential_genes_list]
print(&#39;Average number of transposons of nonessential genes: %.0f &#39; % np.nanmean(tn_pernonessentialgene_list))
print(&#39;Median number of transposons of nonessential genes: %.0f &#39; % np.nanmedian(tn_pernonessentialgene_list))
print(&#39;Standard deviation number of transposons of nonessential genes: %.0f &#39; % np.nanstd(tn_pernonessentialgene_list))
print(&#39;Maximum number of transposons of nonessential genes: %.0f &#39; % np.nanmax(tn_pernonessentialgene_list))
print(&#39;Minimum number of transposons of nonessential genes: %.0f &#39; % np.nanmin(tn_pernonessentialgene_list))
print(&#39;&#39;)
</pre></div>
</div>
</div>
<div class="section" id="create-a-histogram">
<h3><span class="section-number">2.1.7. </span>Create a histogram<a class="headerlink" href="#create-a-histogram" title="Permalink to this headline">¶</a></h3>
<p>A histogram is plotted where the number of reads and number of transposons are shown for essential and non-essential genes together.
This shows the distribution of the number of reads and insertions for the complete dataset.</p>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>plt.figure(figsize=(15,5))
grid = plt.GridSpec(1, 2, wspace=0.4, hspace=0.3)

ax1 = plt.subplot(grid[0,0])
n_reads,bins_reads,patches_reads = ax1.hist(reads_pergene_list, density=False, bins=range(min(reads_pergene_list), max(reads_pergene_list) + 1000, 1000), facecolor=&#39;blue&#39;,alpha=0.5, label=&#39;Reads&#39;)
ax1.axvline(x=np.nanmedian(reads_pergene_list), linestyle=&#39;--&#39;, label=&#39;Median Reads&#39;, color=&#39;k&#39;)
ax1.set_xlim([0,50000])
ax1.set_xlabel(&#39;Reads per gene (all genes)&#39;)
ax1.set_ylabel(&#39;Occurance&#39;)
ax1.legend()
ax1.grid()

ax2 = plt.subplot(grid[0,1])
n_tn,bins_tn,patches_tn = ax2.hist(tn_pergene_list, density=False, bins=range(min(tn_pergene_list), max(tn_pergene_list) + 50, 50), facecolor=&#39;red&#39;,alpha=0.5, label=&#39;Transposons&#39;)
ax2.axvline(x=np.nanmedian(tn_pergene_list), linestyle=&#39;--&#39;, label=&#39;Median Transposons&#39;, color=&#39;k&#39;)
ax2.set_xlim([0,1000])
ax2.set_xlabel(&#39;Counts per gene (all genes)&#39;)
ax2.set_ylabel(&#39;Occurance&#39;)
ax2.legend()
ax2.grid()
plt.show()
</pre></div>
</div>
</div>
<div class="section" id="create-a-violin-plot">
<h3><span class="section-number">2.1.8. </span>Create a violin plot<a class="headerlink" href="#create-a-violin-plot" title="Permalink to this headline">¶</a></h3>
<p>This shows the distribution for the reads and insertions for the dataset separated by annotated essentiality.</p>
<p>When <code class="docutils literal notranslate"><span class="pre">normalization</span></code> is set to <code class="docutils literal notranslate"><span class="pre">True</span></code>, the plots represent read- and transposon densities. Otherwise the plots show the actual counts.</p>
<div class="highlight-{code-cell}ipython3 notranslate"><div class="highlight"><pre><span></span>plt.figure(figsize=(15,5))
grid = plt.GridSpec(1, 2, wspace=0.4, hspace=0.3)


ax1 = plt.subplot(grid[0,0])
sns.set(style=&#39;whitegrid&#39;, palette=&#39;pastel&#39;, color_codes=True)
df[&#39;Reads&#39;] = &#39;&#39;
if normalize == &#39;False&#39;:
    sns.violinplot(x=&#39;Reads&#39;,y=&quot;Reads_per_gene&quot;, hue=&#39;Essential_gene&#39;, inner=&#39;quarter&#39;, gridsize=1000, palette={False:&#39;r&#39;, True:&#39;g&#39;}, split=True, cut=0, orient=&#39;v&#39;, data=df, ax=ax1)
else:
    sns.violinplot(x=&#39;Reads&#39;,y=&quot;Read_density_per_gene&quot;, hue=&#39;Essential_gene&#39;, inner=&#39;quarter&#39;, gridsize=1000, palette={False:&#39;r&#39;, True:&#39;g&#39;}, split=True, cut=0, orient=&#39;v&#39;, data=df, ax=ax1)
ax1.set_ylim([0,25])
ax1.grid()


ax2 = plt.subplot(grid[0,1])
df[&#39;Transposons&#39;] = &#39;&#39;
if normalize == False:
    sns.violinplot(x=&#39;Transposons&#39;,y=&quot;Transposons_per_gene&quot;, hue=&#39;Essential_gene&#39;, inner=&#39;quarter&#39;, gridsize=1000, palette={&#39;False&#39;:&#39;r&#39;, &#39;True&#39;:&#39;g&#39;}, split=True, cut=0, orient=&#39;v&#39;, data=df, ax=ax2)
else:
    sns.violinplot(x=&#39;Transposons&#39;,y=&quot;Transposon_density_per_gene&quot;, hue=&#39;Essential_gene&#39;, inner=&#39;quarter&#39;, gridsize=1000, palette={False:&#39;r&#39;, True:&#39;g&#39;}, split=True, cut=0, orient=&#39;v&#39;, data=df, ax=ax2)
ax2.set_ylim([0.0,0.5])
ax2.grid()

plt.show()
</pre></div>
</div>
</div>
</div>
<div class="section" id="interpretation">
<h2><span class="section-number">2.2. </span>Interpretation<a class="headerlink" href="#interpretation" title="Permalink to this headline">¶</a></h2>
<p>Initially it is expected that the number of reads and the number of insertions are higher for the non-essential genes as these are more tolerant to transposons than the essential genes.
Note that ‘essential’ means ‘annotated essential’ here, and not ‘essential based on the data’.
It might occur that genes are annotated as (non-)essential, but that this does not agree with the data.
This might skew the data shown in the graphs, but since it is expected that these situations do not occur frequently these graphs can still give a good indication about the differences in number of reads and insertions per (non-)essential gene.</p>
<p>Ideally the median values of the essential and nonessential genes are far apart from each other, so that a clear distinction can be made between which genes are essential and which are not based on the transposon counts. If this is not the case or when there is significant overlap between the distribution, no definite conclusion can be drawn about the essentiality.</p>
<p>Usually, essential genes have little transposon counts, but nonessential genes can have both high or low transposon counts (for an explanation on this effect, see the paper by Michel <a class="reference external" href="http://et.al">et.al</a>. 2017).
This means that if a high transposon count occurs, the gene is probably nonessential.
But if a low transposon count is present, typically no definite conclusion can be drawn about its essentiality.</p>
</div>
<div class="section" id="bibliography">
<h2><span class="section-number">2.3. </span>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Michel, A. H., Hatakeyama, R., Kimmig, P., Arter, M., Peter, M., Matos, J., … &amp; Kornmann, B. (2017). Functional mapping of yeast genomes by saturated transposition. Elife, 6, e23570.</p></li>
<li><p>Segal, E. S., Gritsenko, V., Levitan, A., Yadav, B., Dror, N., Steenwyk, J. L., … &amp; Kunze, R. (2018). Gene essentiality analyzed by in vivo transposon mutagenesis and machine learning in a stable haploid isolate of Candida albicans. MBio, 9(5), e02048-18.</p></li>
</ul>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="gene_reads.html" title="previous page"><span class="section-number">1. </span>Gene reads script</a>
    <a class='right-next' id="next-link" href="statistics_perchromosome_Compare.html" title="next page"><span class="section-number">3. </span>Comparison of the statistics per chromosome</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Gregory van Beek<br/>
        
            &copy; Copyright GPL License 3.0.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>