

<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>3. Comparison of the statistics per chromosome &#8212; SATAY DATA ANALYSIS LAAN-LAB TU DELFT</title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css" integrity="sha384-KA6wR/X5RY4zFAHpv/CnoG2UW1uogYfdnP67Uv7eULvTveboZJg0qUpmJZb5VqzN" crossorigin="anonymous">
    <link href="_static/css/index.css" rel="stylesheet">
    <link rel="stylesheet" href="_static/sphinx-book-theme.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/jupyter-sphinx.css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/sphinx-book-theme.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script src="_static/togglebutton.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/mystnb.js"></script>
    <script src="_static/sphinx-book-theme.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="_static/Greg.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="2. Statistics per gene" href="statistics_pergene.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="docsearch:language" content="en">



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="index.html">
  
  <img src="_static/Greg.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">SATAY DATA ANALYSIS LAAN-LAB TU DELFT</h1>
  
</a>
</div>

<form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>

<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="Introduction.html">
   To reproduce an example of  Jupyter book, and you are a Windows user:
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Get started
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="Intro/satay_analysis_notes.html">
   1. All what you need to know from SATAY
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Intro/data-analysis-workflow.html">
   2. Data analysis steps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="MysT_Notebooks.html">
   3. Notebooks written entirely in markdown
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Transposon reads from raw data
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="transposonmapping_satay.html">
   1. Python Translation of the Matlab code from Kornmann Lab
  </a>
 </li>
</ul>
<p class="caption">
 <span class="caption-text">
  Postprocessing scripts
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="gene_reads.html">
   1. Gene reads script
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="statistics_pergene.html">
   2. Statistics per gene
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   3. Comparison of the statistics per chromosome
  </a>
 </li>
</ul>

</nav>

 <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        <div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    
    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="_sources/statistics_perchromosome_Compare.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/statistics_perchromosome_Compare.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
    
</div>
        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/leilaicruz/SATAY-jupyter-book"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/leilaicruz/SATAY-jupyter-book/issues/new?title=Issue%20on%20page%20%2Fstatistics_perchromosome_Compare.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org//v2/gh/leilaicruz/SATAY-jupyter-book/gh-pages?urlpath=lab/tree/statistics_perchromosome_Compare.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   3.1. Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#the-python-script">
   3.2. The python script
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     3.2.1. Input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#loading-files">
     3.2.2. Loading files
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#set-name-of-text-file-and-create-header">
     3.2.3. Set name of text file and create header
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#get-chromosome-information">
     3.2.4. Get chromosome information
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#create-lists-for-the-variables">
     3.2.5. Create lists for the variables
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#determine-statistics">
     3.2.6. Determine statistics
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#creating-text-file">
     3.2.7. Creating text file
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#printing-result">
     3.2.8. Printing result
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#interpretation">
   3.3. Interpretation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bibliography">
   3.4. Bibliography
  </a>
 </li>
</ul>

        </nav>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="comparison-of-the-statistics-per-chromosome">
<h1><span class="section-number">3. </span>Comparison of the statistics per chromosome<a class="headerlink" href="#comparison-of-the-statistics-per-chromosome" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2><span class="section-number">3.1. </span>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This script gives a numerical overview of the two datasets in the form of a textfile.
This can be used for comparing statistical values of two datasets.</p>
</div>
<div class="section" id="the-python-script">
<h2><span class="section-number">3.2. </span>The python script<a class="headerlink" href="#the-python-script" title="Permalink to this headline">¶</a></h2>
<p>The script inputs two .bed files in a list that were generated using the Matlab code provided by the Kornmann lab and the path and name of a text file where the results will be written.
Currently 11 different statistical values are determined, but this can be easily extended (see explainatory text between the codes below).</p>
<ol class="simple">
<li><p>Number of transposon insertions.</p></li>
<li><p>Percentage of the chromosome that is covered by transposons</p></li>
<li><p>Mean distance between transposon insertions.</p></li>
<li><p>Median distance between transposon insertions.</p></li>
<li><p>25th percentile of the distance between transposon insertions.</p></li>
<li><p>75th percentile of the distance between transposon insertions.</p></li>
<li><p>Largest area devoid of transposons</p></li>
<li><p>Mean number of reads per transposon</p></li>
<li><p>Median numbr of reads per transposon</p></li>
<li><p>25th percentile reads per transposon</p></li>
<li><p>75th percentile reads per transposon</p></li>
</ol>
<div class="section" id="input">
<h3><span class="section-number">3.2.1. </span>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>The function inputs either a gene name (<code class="docutils literal notranslate"><span class="pre">gene_name</span></code>, type=string) or a region (<code class="docutils literal notranslate"><span class="pre">region</span></code>, type=list) specified as a list with three entries (chromosome number as a roman numeral, start position of the region and the end position respectively).
The variable <code class="docutils literal notranslate"><span class="pre">gene_name</span></code> can be set to any gene name or <code class="docutils literal notranslate"><span class="pre">holocus</span></code> or <code class="docutils literal notranslate"><span class="pre">ho-locus</span></code>.
Next it requires the bed-file (<code class="docutils literal notranslate"><span class="pre">bed_file</span></code>, type=string) which is created by the Matlab code provided from <a class="reference external" href="https://sites.google.com/site/satayusers/complete-protocol/bioinformatics-analysis/matlab-script">the Kornman lab</a>.
Finally, the figure can be automatically saved (at a location specified in the beginning of the function) by setting <code class="docutils literal notranslate"><span class="pre">savefigure</span></code> to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<p>The custom build functions (stored in the <a class="reference external" href="https://github.com/Gregory94/LaanLab-SATAY-DataAnalysis/tree/master/python_modules">‘python modules’ folder on Github</a>) that are required are:</p>
<ul class="simple">
<li><p>chromosome_and_gene_positions.chromosome_position</p></li>
<li><p>chromosome_and_gene_positions.chromosome_roman_to_arabic</p></li>
<li><p>chromosome_names_in_fies.chromosome_name_bedfile</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">tabulate</span> <span class="kn">import</span> <span class="n">tabulate</span>

<span class="n">file_dirname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="s1">&#39;__file__&#39;</span><span class="p">)</span>
<span class="c1"># sys.path.insert(1,os.path.join(file_dirname,&#39;python_modules&#39;))</span>
<span class="kn">from</span> <span class="nn">Python_scripts.python_modules.chromosome_and_gene_positions</span> <span class="kn">import</span> <span class="n">chromosome_position</span><span class="p">,</span> <span class="n">chromosomename_roman_to_arabic</span>
<span class="kn">from</span> <span class="nn">Python_scripts.python_modules.chromosome_names_in_files</span> <span class="kn">import</span> <span class="n">chromosome_name_bedfile</span>


<span class="c1">#FUNCTION INPUTS</span>
<span class="n">bed_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;satay_analysis_testdata&#39;</span><span class="p">,</span><span class="s1">&#39;Output_Processing&#39;</span><span class="p">,</span><span class="s1">&#39;Cerevisiae_WT2_Michel2017_trimmed1.bam.bed&#39;</span><span class="p">),</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;satay_analysis_testdata&#39;</span><span class="p">,</span><span class="s1">&#39;Output_Processing&#39;</span><span class="p">,</span><span class="s1">&#39;Cerevisiae_WT2_Michel2017_trimmed2.bam.bed&#39;</span><span class="p">)]</span>
<span class="n">text_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;satay_analysis_testdata&#39;</span><span class="p">,</span><span class="s1">&#39;Output_Processing&#39;</span><span class="p">,</span><span class="s1">&#39;Cerevisiae_WT2_Michel2017_trimmed1-2_Compare.txt&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">test_path</span> <span class="ow">in</span> <span class="n">bed_files</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">test_path</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>True
True
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="loading-files">
<h3><span class="section-number">3.2.2. </span>Loading files<a class="headerlink" href="#loading-files" title="Permalink to this headline">¶</a></h3>
<p>Next additional files are loaded. Change this to your local paths leading to the gff-file (for example downloaded from SGD [https://www.yeastgenome.org/] or get a copy from the docs folder on Github [https://github.com/Gregory94/LaanLab-SATAY-DataAnalysis]).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gff_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;Python_scripts&#39;</span><span class="p">,</span><span class="s1">&#39;Data_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Saccharomyces_cerevisiae.R64-1-1.99.gff3&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="set-name-of-text-file-and-create-header">
<h3><span class="section-number">3.2.3. </span>Set name of text file and create header<a class="headerlink" href="#set-name-of-text-file-and-create-header" title="Permalink to this headline">¶</a></h3>
<p>The path and name that are given as input for saving the text file is extended with ‘DataCompare.txt’ to make sure the file is recognizable and have the right extension.</p>
<p>Secondly, the header for the text file is created here.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">text_file</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
    <span class="n">text_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">bed_files</span><span class="p">[</span><span class="mi">0</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_DataCompare.txt&#39;</span>

<span class="k">if</span> <span class="n">text_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">t</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">text_file</span><span class="p">,</span> <span class="s1">&#39;w+&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">bed_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Statistical values for datafile: &#39;</span> <span class="o">+</span> <span class="n">bed_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">bed_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Statistical values for datafile: &#39;</span> <span class="o">+</span> <span class="n">bed_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; and &#39;</span> <span class="o">+</span> <span class="n">bed_files</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">t</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="get-chromosome-information">
<h3><span class="section-number">3.2.4. </span>Get chromosome information<a class="headerlink" href="#get-chromosome-information" title="Permalink to this headline">¶</a></h3>
<p>Determine the lengths and position of the different chromosomes and get a list of the chromosome names in terms of roman numerals.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr_length_dict</span><span class="p">,</span> <span class="n">chr_start_pos_dict</span><span class="p">,</span> <span class="n">chr_end_pos_dict</span> <span class="o">=</span> <span class="n">chromosome_position</span><span class="p">(</span><span class="n">gff_file</span><span class="p">)</span>

<span class="n">roman_to_arabic_dict</span> <span class="o">=</span> <span class="n">chromosomename_roman_to_arabic</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">chromosome_romannames_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">roman</span> <span class="ow">in</span> <span class="n">roman_to_arabic_dict</span><span class="p">:</span>
    <span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">roman</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="create-lists-for-the-variables">
<h3><span class="section-number">3.2.5. </span>Create lists for the variables<a class="headerlink" href="#create-lists-for-the-variables" title="Permalink to this headline">¶</a></h3>
<p>All the variables that are determined are put in individual lists.
When new statistics needs to be determined, add a new list for each new value.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">N_Tn_Ins</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Cov_Percent</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Mean_Tn_Dist</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Median_Tn_Dist</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Tn_Dist_25th_Percent</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Tn_Dist_75th_Percent</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Tn_Gap</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Mean_Reads_PerTn</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Median_Reads_PerTn</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Reads_25th_percent</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">Reads_75th_percent</span> <span class="o">=</span> <span class="p">[]</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="determine-statistics">
<h3><span class="section-number">3.2.6. </span>Determine statistics<a class="headerlink" href="#determine-statistics" title="Permalink to this headline">¶</a></h3>
<p>For each chromosome in each .bed file the statistical values are determined and stored in the respective lists.
The values are also determined for the entire genome.</p>
<p>First a for loop is done over all bed files and for each bed file the names of the chromosomes as they are stored in the bed files are determined.
Next, some statistical values are determined and depending whether it is first file written a text file is generated or the values are appended to the already exsiting text file.
After all chromosomes the statistics are determined for the entire genome as well.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bed_file_counter</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">bed_file</span> <span class="ow">in</span> <span class="n">bed_files</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">bed_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>

    <span class="n">chrom_names_dict</span><span class="p">,</span> <span class="n">chrom_start_index_dict</span><span class="p">,</span> <span class="n">chrom_end_index_dict</span> <span class="o">=</span> <span class="n">chromosome_name_bedfile</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>
    <span class="n">chrom_loop</span> <span class="o">=</span> <span class="n">chrom_names_dict</span>

<span class="c1">#DETERMINE STATISTICS</span>
    <span class="n">bp_between_tn_insertions_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">reads_per_tn_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chrom_loop</span><span class="p">:</span>
        <span class="n">tn_insertion_position_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">reads_per_tn_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">[</span><span class="n">chrom_start_index_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">):</span><span class="n">chrom_end_index_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="n">tn_insertion_position_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
            <span class="n">reads_per_tn_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span><span class="o">-</span><span class="mi">100</span><span class="p">)</span><span class="o">/</span><span class="mi">20</span><span class="p">)</span>
        <span class="n">bp_between_tn_insertions</span> <span class="o">=</span> <span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">y</span><span class="o">-</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">tn_insertion_position_list</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">tn_insertion_position_list</span><span class="p">[</span><span class="mi">1</span><span class="p">:])]</span>
        <span class="n">bp_between_tn_insertions</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">tn_insertion_position_list</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="c1">#ADD START OF GENE (bp=0)</span>
        <span class="n">bp_between_tn_insertions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">chr_length_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="o">-</span> <span class="n">tn_insertion_position_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="c1">#ADD END OF GENE (bp=GENE_LENGTH-TRANSPOSON INSERTION LAST GENE)</span>
        <span class="n">bp_between_tn_insertions_dict</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">bp_between_tn_insertions</span>
        <span class="n">reads_per_tn_dict</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">reads_per_tn_list</span>

        <span class="n">tn_insertion_meanfrequency</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">)</span>
        <span class="n">tn_insertion_25percentilefrequency</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">,</span><span class="mi">25</span><span class="p">)</span>
        <span class="n">tn_insertion_medianfrequency</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmedian</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">)</span>
        <span class="n">tn_insertion_75percentilefrequency</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">,</span><span class="mi">75</span><span class="p">)</span>

<span class="c1">#IF IT IS THE FIRST BED FILE, CREATE TITLES FOR EACH STATISTIC TO WRITE IN THE TEXT FILE.</span>
        <span class="k">if</span> <span class="n">bed_file_counter</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Print information chromosome &#39;</span> <span class="o">+</span> <span class="n">chrom</span> <span class="o">+</span> <span class="s1">&#39; with length &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">chr_length_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">)))</span>
            <span class="n">N_Tn_Ins</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;Number of transposon insertions&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            <span class="n">Cov_Percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;Coverage percent&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">tn_insertion_position_list</span><span class="p">)</span><span class="o">/</span><span class="n">chr_length_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            
            <span class="n">Mean_Tn_Dist</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;Mean distance between insertions&#39;</span><span class="p">,</span> <span class="n">tn_insertion_meanfrequency</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            <span class="n">Median_Tn_Dist</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;Median distance between insertions&#39;</span><span class="p">,</span> <span class="n">tn_insertion_medianfrequency</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            <span class="n">Tn_Dist_25th_Percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;25th percentile distance between insertions&#39;</span><span class="p">,</span> <span class="n">tn_insertion_25percentilefrequency</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            <span class="n">Tn_Dist_75th_Percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;75th percentile distance between insertions&#39;</span><span class="p">,</span> <span class="n">tn_insertion_75percentilefrequency</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            
            <span class="n">Tn_Gap</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;Largest area devoid of transposons&#39;</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            
            <span class="n">Mean_Reads_PerTn</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;Mean number of reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            <span class="n">Median_Reads_PerTn</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;Median number of reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmedian</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            <span class="n">Reads_25th_percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;25th percentile reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">,</span><span class="mi">25</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
            <span class="n">Reads_75th_percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;75th percentile reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">,</span><span class="mi">75</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>

<span class="c1">#IF IT IS NOT THE FIRST BED FILE, APPEND THE STATISTICS FOR THE CURRENT BED FILE TO THE EXISTING STATISTICS OF THE PREVIOUS BED FILE.</span>
        <span class="k">elif</span> <span class="n">bed_file_counter</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">N_Tn_Ins</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">)</span>
            <span class="n">Cov_Percent</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tn_insertion_position_list</span><span class="p">)</span><span class="o">/</span><span class="n">chr_length_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span>
            
            <span class="n">Mean_Tn_Dist</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">tn_insertion_meanfrequency</span>
            <span class="n">Median_Tn_Dist</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">tn_insertion_medianfrequency</span>
            <span class="n">Tn_Dist_25th_Percent</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">tn_insertion_25percentilefrequency</span>
            <span class="n">Tn_Dist_75th_Percent</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">tn_insertion_75percentilefrequency</span>
            
            <span class="n">Tn_Gap</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">bp_between_tn_insertions</span><span class="p">)</span>
            
            <span class="n">Mean_Reads_PerTn</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">)</span>
            <span class="n">Median_Reads_PerTn</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmedian</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">)</span>
            <span class="n">Reads_25th_percent</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span>  <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">,</span><span class="mi">25</span><span class="p">)</span>
            <span class="n">Reads_75th_percent</span><span class="p">[</span><span class="n">chromosome_romannames_list</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">chrom</span><span class="p">)][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_list</span><span class="p">,</span><span class="mi">75</span><span class="p">)</span>

<span class="c1">#DETERMINE STATISTICS FOR THE ENTIRE GENOME</span>
    <span class="n">bp_between_tn_insertions_genome</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">number_tn_insertions_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">reads_per_tn_genome</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">number_tn_insertions_genome</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chrom_loop</span><span class="p">:</span>
        <span class="n">number_tn_insertions_genome</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads_per_tn_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">bp_between</span> <span class="ow">in</span> <span class="n">bp_between_tn_insertions_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">):</span>
            <span class="n">bp_between_tn_insertions_genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bp_between</span><span class="p">)</span>
        <span class="n">number_tn_insertions_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bp_between_tn_insertions_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">)))</span>
        <span class="k">for</span> <span class="n">reads_tn</span> <span class="ow">in</span> <span class="n">reads_per_tn_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">chrom</span><span class="p">):</span>
            <span class="n">reads_per_tn_genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reads_tn</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">bed_file_counter</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">N_Tn_Ins</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span><span class="s1">&#39;Number of insertions&#39;</span><span class="p">,</span><span class="n">number_tn_insertions_genome</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Cov_Percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;Coverage percent&#39;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">number_tn_insertions_list</span><span class="p">)</span><span class="o">/</span><span class="nb">sum</span><span class="p">(</span><span class="n">chr_length_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Mean_Tn_Dist</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;Mean distance between insertions&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Median_Tn_Dist</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;Median distance between insertions&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmedian</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Tn_Dist_25th_Percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;25th percentile distance between insertions&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">,</span><span class="mi">25</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Tn_Dist_75th_Percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;75th percentile distance between insertions&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">,</span><span class="mi">75</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Tn_Gap</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Mean_Reads_PerTn</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;Mean number of reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Median_Reads_PerTn</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;Median number of reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmedian</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Reads_25th_percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;25th percentile reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">,</span><span class="mi">25</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>
        <span class="n">Reads_75th_percent</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;75th percentile reads per transposon&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">,</span><span class="mi">75</span><span class="p">),</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>

    <span class="k">elif</span> <span class="n">bed_file_counter</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">N_Tn_Ins</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">number_tn_insertions_genome</span>
        <span class="n">Cov_Percent</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">number_tn_insertions_list</span><span class="p">)</span><span class="o">/</span><span class="nb">sum</span><span class="p">(</span><span class="n">chr_length_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span><span class="o">*</span><span class="mi">100</span>
        <span class="n">Mean_Tn_Dist</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">)</span>
        <span class="n">Median_Tn_Dist</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmedian</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">)</span>
        <span class="n">Tn_Dist_25th_Percent</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">,</span><span class="mi">25</span><span class="p">)</span>
        <span class="n">Tn_Dist_75th_Percent</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">bp_between_tn_insertions_genome</span><span class="p">,</span><span class="mi">75</span><span class="p">)</span>
        <span class="n">Tn_Gap</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="n">Mean_Reads_PerTn</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">)</span>
        <span class="n">Median_Reads_PerTn</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmedian</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">)</span>
        <span class="n">Reads_25th_percent</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">,</span><span class="mi">25</span><span class="p">)</span>
        <span class="n">Reads_75th_percent</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">reads_per_tn_genome</span><span class="p">,</span><span class="mi">75</span><span class="p">)</span>

    <span class="n">bed_file_counter</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Print information chromosome I with length 230218
Print information chromosome II with length 813184
Print information chromosome III with length 316620
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Print information chromosome IV with length 1531933
Print information chromosome V with length 576874
Print information chromosome VI with length 270161
Print information chromosome VII with length 1090940
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Print information chromosome VIII with length 562643
Print information chromosome IX with length 439888
Print information chromosome X with length 745751
Print information chromosome XI with length 666816
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Print information chromosome XII with length 1078177
Print information chromosome XIII with length 924431
Print information chromosome XIV with length 784333
Print information chromosome XV with length 1091291
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Print information chromosome XVI with length 948066
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="creating-text-file">
<h3><span class="section-number">3.2.7. </span>Creating text file<a class="headerlink" href="#creating-text-file" title="Permalink to this headline">¶</a></h3>
<p>Write the stored statistical value to the text file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Writing to text file...&#39;</span><span class="p">)</span>

<span class="n">header0</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">,</span><span class="s1">&#39;item&#39;</span><span class="p">,</span><span class="s1">&#39;Dataset 1&#39;</span><span class="p">,</span><span class="s1">&#39;Dataset 2&#39;</span><span class="p">]</span>
<span class="n">header</span>  <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;          &#39;</span><span class="p">,</span><span class="s1">&#39;    &#39;</span><span class="p">,</span><span class="s1">&#39;         &#39;</span><span class="p">,</span><span class="s1">&#39;         &#39;</span><span class="p">]</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">text_file</span><span class="p">,</span><span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">N_Tn_Ins</span><span class="p">)):</span>
        <span class="n">table</span> <span class="o">=</span> <span class="p">[</span><span class="n">N_Tn_Ins</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Cov_Percent</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Mean_Tn_Dist</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Median_Tn_Dist</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Tn_Dist_25th_Percent</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">Tn_Dist_75th_Percent</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Tn_Gap</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Mean_Reads_PerTn</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Median_Reads_PerTn</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                    <span class="n">Reads_25th_percent</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">Reads_75th_percent</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tabulate</span><span class="p">(</span><span class="n">table</span><span class="p">,</span><span class="n">tablefmt</span><span class="o">=</span><span class="s1">&#39;github&#39;</span><span class="p">,</span><span class="n">headers</span><span class="o">=</span><span class="n">header0</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tabulate</span><span class="p">(</span><span class="n">table</span><span class="p">,</span><span class="n">tablefmt</span><span class="o">=</span><span class="s1">&#39;github&#39;</span><span class="p">,</span><span class="n">headers</span><span class="o">=</span><span class="n">header</span><span class="p">))</span>
        <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;...Writing complete.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Writing to text file...
...Writing complete.
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="printing-result">
<h3><span class="section-number">3.2.8. </span>Printing result<a class="headerlink" href="#printing-result" title="Permalink to this headline">¶</a></h3>
<p>Showing part of the text file to give an example of the created text file.
(This is just for showing the result in this notebook and is not important for the code itself.
Can be removed).</p>
<div class="cell tag_outputPrepend docutils container">
<div class="cell_input docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">text_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">t</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">20</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">readline</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>Statistical values for datafile: satay_analysis_testdata/Output_Processing/Cerevisiae_WT2_Michel2017_trimmed1.bam.bed and satay_analysis_testdata/Output_Processing/Cerevisiae_WT2_Michel2017_trimmed2.bam.bed

| chromosome   | item                                        |   Dataset 1 |   Dataset 2 |

|--------------|---------------------------------------------|-------------|-------------|

| I            | Number of transposon insertions             | 16437       | 16342       |

| I            | Coverage percent                            |     7.13975 |     7.09849 |

| I            | Mean distance between insertions            |    14.0052  |    14.0866  |

| I            | Median distance between insertions          |     5       |     5       |

| I            | 25th percentile distance between insertions |     2       |     2       |

| I            | 75th percentile distance between insertions |    13       |    13       |

| I            | Largest area devoid of transposons          |  1142       |  1142       |

| I            | Mean number of reads per transposon         |    30.6532  |    30.8216  |

| I            | Median number of reads per transposon       |     9       |     9       |

| I            | 25th percentile reads per transposon        |     2       |     2       |

| I            | 75th percentile reads per transposon        |    30       |    30       |

|              |                                             |             |             |

|--------------|---------------------------------------------|-------------|-------------|

| II           | Number of transposon insertions             | 35907       | 35767       |

| II           | Coverage percent                            |     4.41561 |     4.39839 |

| II           | Mean distance between insertions            |    22.6463  |    22.735   |

| II           | Median distance between insertions          |     7       |     7       |

</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="interpretation">
<h2><span class="section-number">3.3. </span>Interpretation<a class="headerlink" href="#interpretation" title="Permalink to this headline">¶</a></h2>
<p>The statistical values that are determined here for two datasets give an indication of some of the properties of the datasets.
Together with the TransposonRead_Profile_Compare.py script, this can be helpful when comparing two datasets with each other to possibly improve the preprocessing steps.
This script can relatively easy be extended with more statistical values in the future.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Coverage</span> <span class="pre">percentage</span></code> is the number of transposons divided by the number of basepairs of the chromosome.
The distance between transposon insertions is determined by the taking the absolute difference between all subsequent transposon and for the first and last transposon the distance is determined from the beginning and until the end of the chromosome, respectively.
The same goes for the median and the percentiles.
The largest distance between subsequent transposons is displayed as the largest area devoid of transposons.
The values regarding the number of reads are directly extracted from the bed files.</p>
</div>
<div class="section" id="bibliography">
<h2><span class="section-number">3.4. </span>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Michel, A. H., Hatakeyama, R., Kimmig, P., Arter, M., Peter, M., Matos, J., … &amp; Kornmann, B. (2017). Functional mapping of yeast genomes by saturated transposition. Elife, 6, e23570.</p></li>
</ul>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="statistics_pergene.html" title="previous page"><span class="section-number">2. </span>Statistics per gene</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Gregory van Beek<br/>
        
            &copy; Copyright GPL License 3.0.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    <script src="_static/js/index.js"></script>
    
  </body>
</html>