
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Volcano plots to two different libraries &#8212; SATAY pipeline at Delft :)</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet" />
  <link href="../_static/css/index.c5995385ac14fb8791e8eb36b4908be2.css" rel="stylesheet" />

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.acff12b8f9c144ce68a297486a2fa670.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.1c5a1a01449ed65a7b51.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/Greg.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Searching for genomic features of transposition events" href="06-genomicfeatures_dataframe.html" />
    <link rel="prev" title="Create a scatterplot for all genes and all essential genes." href="04-scatterplot_genes.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      <img src="../_static/genetica.jpg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">SATAY pipeline at Delft :)</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../documentation_satay.html">
   Documentation for processing SATAY
   <!-- omit in toc -->
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Introduction.html">
   Status of the project
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../00-Intro/data-analysis-workflow.html">
   Data analysis steps
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../00-Intro/flowchart.html">
     Flowchart of the pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00-Intro/Installation_Guide_SATAY_Analysis_Software.html">
     Software installation guide for whole genome sequencing and transposon mapping
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../00-Intro/satay_analysis_notes.html">
     All what you need to know from SATAY
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../01-bash-GUI/processing_workflow.html">
   Code for the bash GUI
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="00-Introduction.html">
   Python scripts
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="01-clean_bedwigfiles.html">
     Cleaning bed and wig files
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="02-transposonread_profileplot_genome.html">
     Transposon profile for the whole genome
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="03-transposonread_profileplot.html">
     Transposon profile for a chromosome of interest
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="04-scatterplot_genes.html">
     Create a scatterplot for all genes and all essential genes.
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Volcano plots to two different libraries
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="06-genomicfeatures_dataframe.html">
     Searching for genomic features  of transposition events
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../03-docker-doc/00-Docker-Users.html">
   Build your own docker container with ours satay pipeline
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/02-postprocessing-scripts/05-volcanoplot.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/SATAY-LL/SATAY-jupyter-book"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/SATAY-LL/SATAY-jupyter-book/issues/new?title=Issue%20on%20page%20%2F02-postprocessing-scripts/05-volcanoplot.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#how-to-use-the-code-in-the-same-script">
   How to use the code in the same script
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="volcano-plots-to-two-different-libraries">
<h1>Volcano plots to two different libraries<a class="headerlink" href="#volcano-plots-to-two-different-libraries" title="Permalink to this headline">¶</a></h1>
<p>This script creates a volcanoplot to show the significance of fold change between two <a class="reference external" href="http://datasets.It">datasets.It</a> is based on this website:
- <a class="reference external" href="https://towardsdatascience.com/inferential-statistics-series-t-test-using-numpy-2718f8f9bf2f">https://towardsdatascience.com/inferential-statistics-series-t-test-using-numpy-2718f8f9bf2f</a>
- <a class="reference external" href="https://www.statisticshowto.com/independent-samples-t-test/">https://www.statisticshowto.com/independent-samples-t-test/</a></p>
<p>Code for showing gene name when hovering over datapoint is based on:
- <a class="reference external" href="https://stackoverflow.com/questions/7908636/possible-to-make-labels-appear-when-hovering-over-a-point-in-matplotlib">https://stackoverflow.com/questions/7908636/possible-to-make-labels-appear-when-hovering-over-a-point-in-matplotlib</a></p>
<p>T-test is measuring the number of standard deviations our measured mean is from the baseline mean, while taking into
account that the standard deviation of the mean can change as we get more data. Look here for more information:</p>
<ul class="simple">
<li><p>Independent T-test : <a class="reference external" href="https://www.statisticshowto.com/independent-samples-t-test/">https://www.statisticshowto.com/independent-samples-t-test/</a></p></li>
</ul>
<div class="tip dropdown admonition">
<p class="admonition-title">Click here to see the code!</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>


<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">file_dirname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">&#39;__file__&#39;</span><span class="p">))</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">file_dirname</span><span class="p">,</span><span class="s1">&#39;python_modules&#39;</span><span class="p">))</span>
<span class="kn">from</span> <span class="nn">dataframe_from_pergene</span> <span class="kn">import</span> <span class="n">dataframe_from_pergenefile</span>




<span class="c1">#%%</span>
<span class="k">def</span> <span class="nf">volcano</span><span class="p">(</span><span class="n">path_a</span><span class="p">,</span> <span class="n">filelist_a</span><span class="p">,</span> <span class="n">path_b</span><span class="p">,</span> <span class="n">filelist_b</span><span class="p">,</span> <span class="n">variable</span><span class="o">=</span><span class="s1">&#39;read_per_gene&#39;</span><span class="p">,</span> <span class="n">significance_threshold</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">trackgene_list</span><span class="o">=</span><span class="p">[],</span> <span class="n">figure_title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    This creates a volcano plot that shows the fold change between two libraries and the corresponding p-values.</span>
<span class="sd">    Input:</span>
<span class="sd">        - path_a, path_b: paths to location of the datafiles for library a and library b</span>
<span class="sd">        - filelist_a, filelist_b: list of the names of the datafiles for library a and library b located in path_a and path_b respectively</span>
<span class="sd">        - variable: tn_per_gene, read_per_gene or Nreadsperinsrt (default=&#39;read_per_gene&#39;)</span>
<span class="sd">        - significance_threshold: Treshold value above which the fold change is regarded significant, only for plotting (default=0.01)</span>
<span class="sd">        - normalize: Whether to normalize variable. If set to True, each gene is normalized based on the total count in each dataset (i.e. each file in filelist_) (default=True)</span>
<span class="sd">        - trackgene_list: Enter a list of gene name(s) which will be highlighted in the plot (e.g. [&#39;cdc42&#39;, &#39;nrp1&#39;]). If empty list, no gene will be tracked. (default=[])</span>

<span class="sd">    Output:</span>
<span class="sd">        - volcano_df: pandas dataframe containing:</span>
<span class="sd">            - gene_names</span>
<span class="sd">            - fold change</span>
<span class="sd">            - t statistic</span>
<span class="sd">            - p value</span>
<span class="sd">            - whether p value is above threshold</span>
<span class="sd">        - volcanoplot with the log2 fold change between the two libraries and the -log10 p-value.</span>

<span class="sd">    Fold change is determined by the mean of dataset b (experimental set) divided by the mean of dataset a (reference set).</span>
<span class="sd">    The datasets can be of different length.</span>
<span class="sd">    P-value is determined based on the student t-test (scipy.stats.ttest_ind).</span>

<span class="sd">    NOTE:</span>
<span class="sd">        The fold change is determined by the ratio between the reference and the experimental dataset.</span>
<span class="sd">        When one of the datasets is 0, this is false results for the fold change.</span>
<span class="sd">        To prevent this, the genes with 0 insertions are set to have 5 insertions, and the genes with 0 reads are set to have 25 reads.</span>
<span class="sd">        These values are determined in dicussion with the Kornmann lab.</span>

<span class="sd">    Dependencies:</span>
<span class="sd">        - numpy</span>
<span class="sd">        - scipy</span>
<span class="sd">        - matplotlib</span>
<span class="sd">        - python_modules/dataframe_from_pergene.py (https://github.com/Gregory94/LaanLab-SATAY-DataAnalysis/blob/master/python_modules/dataframe_from_pergene.py)</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">datafiles_list_a</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">datafiles_list_b</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">filelist_a</span><span class="p">:</span>
        <span class="n">datafile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path_a</span><span class="p">,</span> <span class="n">files</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">datafile</span><span class="p">),</span> <span class="s1">&#39;File not found at: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">datafile</span>
        <span class="n">datafiles_list_a</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">filelist_b</span><span class="p">:</span>
        <span class="n">datafile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path_b</span><span class="p">,</span> <span class="n">files</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">datafile</span><span class="p">),</span> <span class="s1">&#39;File not found at: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">datafile</span>
        <span class="n">datafiles_list_b</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">datafile</span><span class="p">)</span>

    <span class="k">del</span> <span class="p">(</span><span class="n">files</span><span class="p">,</span> <span class="n">datafile</span><span class="p">,</span> <span class="n">path_a</span><span class="p">,</span> <span class="n">path_b</span><span class="p">,</span> <span class="n">filelist_a</span><span class="p">,</span> <span class="n">filelist_b</span><span class="p">)</span>



<span class="c1">#%% Extract information from datasets</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Plotting: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">variable</span><span class="p">)</span>
    <span class="n">tn_per_gene_zeroreplace</span> <span class="o">=</span> <span class="mi">5</span> <span class="c1">#Add 5 insertions to every gene</span>
    <span class="n">read_per_gene_zeroreplace</span> <span class="o">=</span> <span class="mi">25</span> <span class="c1">#Add 25 reads to every gene</span>
    <span class="c1"># norm_a = 0</span>
    <span class="c1"># norm_b = 0</span>
    <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">datafile_a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">datafiles_list_a</span><span class="p">):</span>
        <span class="n">tnread_gene_a</span> <span class="o">=</span> <span class="n">dataframe_from_pergenefile</span><span class="p">(</span><span class="n">datafile_a</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">variable</span> <span class="o">==</span> <span class="s1">&#39;tn_per_gene&#39;</span><span class="p">:</span>
                <span class="n">norm_a</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">tn_per_gene</span><span class="p">)</span><span class="c1">#*10**-4</span>
            <span class="k">elif</span> <span class="n">variable</span> <span class="o">==</span> <span class="s1">&#39;read_per_gene&#39;</span><span class="p">:</span>
                <span class="n">norm_a</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">read_per_gene</span><span class="p">)</span><span class="c1">#*10**-7</span>
            <span class="k">elif</span> <span class="n">variable</span> <span class="o">==</span> <span class="s1">&#39;Nreadsperinsrt&#39;</span><span class="p">:</span>
                <span class="n">norm_a</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">read_per_gene</span><span class="p">)</span>

        <span class="c1">#ADD A CONSTANT TO ALL VALUES TO PREVENT A ZERO DIVISION WHEN DETERMINING THE FOLD CHANGE.</span>
        <span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">tn_per_gene</span> <span class="o">=</span> <span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">tn_per_gene</span> <span class="o">+</span> <span class="n">tn_per_gene_zeroreplace</span>
        <span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">read_per_gene</span> <span class="o">=</span> <span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">read_per_gene</span> <span class="o">+</span> <span class="n">read_per_gene_zeroreplace</span>
        <span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">Nreadsperinsrt</span> <span class="o">=</span> <span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">Nreadsperinsrt</span> <span class="o">+</span> <span class="p">(</span><span class="n">read_per_gene_zeroreplace</span><span class="o">/</span><span class="n">tn_per_gene_zeroreplace</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">volcano_df</span> <span class="o">=</span> <span class="n">tnread_gene_a</span><span class="p">[[</span><span class="s1">&#39;gene_names&#39;</span><span class="p">]]</span> <span class="c1">#initialize new dataframe with gene_names</span>
            <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">variable_a_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">tnread_gene_a</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">norm_a</span><span class="p">)</span> <span class="c1">#create numpy array to store normalized data</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variable_a_array</span> <span class="o">=</span> <span class="n">tnread_gene_a</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span> <span class="c1">#create numpy array to store raw data</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">variable_a_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">tnread_gene_a</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">norm_a</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#append normalized data</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variable_a_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">,</span> <span class="n">tnread_gene_a</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#append raw data</span>


    <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">datafile_b</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">datafiles_list_b</span><span class="p">):</span>
        <span class="n">tnread_gene_b</span> <span class="o">=</span> <span class="n">dataframe_from_pergenefile</span><span class="p">(</span><span class="n">datafile_b</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">variable</span> <span class="o">==</span> <span class="s1">&#39;tn_per_gene&#39;</span><span class="p">:</span>
                <span class="n">norm_b</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">tn_per_gene</span><span class="p">)</span><span class="c1">#*10**-4</span>
            <span class="k">elif</span> <span class="n">variable</span> <span class="o">==</span> <span class="s1">&#39;read_per_gene&#39;</span><span class="p">:</span>
                <span class="n">norm_b</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">read_per_gene</span><span class="p">)</span><span class="c1">#*10**-7</span>
            <span class="k">elif</span> <span class="n">variable</span> <span class="o">==</span> <span class="s1">&#39;Nreadsperinsrt&#39;</span><span class="p">:</span>
                <span class="n">norm_b</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">read_per_gene</span><span class="p">)</span>

        <span class="c1">#ADD A CONSTANT TO ALL VALUES TO PREVENT A ZERO DIVISION WHEN DETERMINING THE FOLD CHANGE.</span>
        <span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">tn_per_gene</span> <span class="o">=</span> <span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">tn_per_gene</span> <span class="o">+</span> <span class="n">tn_per_gene_zeroreplace</span>
        <span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">read_per_gene</span> <span class="o">=</span> <span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">read_per_gene</span> <span class="o">+</span> <span class="n">read_per_gene_zeroreplace</span>
        <span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">Nreadsperinsrt</span> <span class="o">=</span> <span class="n">tnread_gene_b</span><span class="o">.</span><span class="n">Nreadsperinsrt</span> <span class="o">+</span> <span class="p">(</span><span class="n">read_per_gene_zeroreplace</span><span class="o">/</span><span class="n">tn_per_gene_zeroreplace</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">variable_b_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">tnread_gene_b</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">norm_b</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variable_b_array</span> <span class="o">=</span> <span class="n">tnread_gene_b</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="n">variable_b_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variable_b_array</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">tnread_gene_b</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">norm_b</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">variable_b_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variable_b_array</span><span class="p">,</span> <span class="n">tnread_gene_b</span><span class="p">[[</span><span class="n">variable</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>


    <span class="k">del</span> <span class="p">(</span><span class="n">datafile_a</span><span class="p">,</span> <span class="n">datafile_b</span><span class="p">,</span> <span class="n">count</span><span class="p">,</span> <span class="n">tnread_gene_b</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">trackgene_list</span> <span class="o">==</span> <span class="p">[]:</span>
        <span class="k">del</span> <span class="n">tnread_gene_a</span>


<span class="c1">#%% APPLY stats.ttest_ind(A,B)</span>
    <span class="n">fc_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">)</span> <span class="c1">#initialize list for storing fold changes</span>
    <span class="n">ttest_tval_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">)</span> <span class="c1">#initialize list for storing t statistics</span>
    <span class="n">ttest_pval_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">)</span> <span class="c1">#initialize list for storing p-values</span>
    <span class="n">signif_thres_list</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">)</span> <span class="c1">#initialize boolean list for indicating datapoints with p-value above threshold</span>

    <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">):</span>

        <span class="n">ttest_val</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">[</span><span class="n">count</span><span class="p">],</span> <span class="n">variable_b_array</span><span class="p">[</span><span class="n">count</span><span class="p">])</span> <span class="c1">#T-test</span>
        <span class="n">ttest_tval_list</span><span class="p">[</span><span class="n">count</span><span class="p">]</span> <span class="o">=</span> <span class="n">ttest_val</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ttest_val</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1">#prevent p=0 to be inputted in log</span>
            <span class="n">ttest_pval_list</span><span class="p">[</span><span class="n">count</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">ttest_val</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ttest_pval_list</span><span class="p">[</span><span class="n">count</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">ttest_pval_list</span><span class="p">[</span><span class="n">count</span><span class="p">]</span> <span class="o">&gt;</span> <span class="o">-</span><span class="mi">1</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">significance_threshold</span><span class="p">):</span>
            <span class="n">signif_thres_list</span><span class="p">[</span><span class="n">count</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1">#DETERMINE FOLD CHANGE PER GENE</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">variable_b_array</span><span class="p">[</span><span class="n">count</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">[</span><span class="n">count</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">fc_list</span><span class="p">[</span><span class="n">count</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">fc_list</span><span class="p">[</span><span class="n">count</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">variable_a_array</span><span class="p">[</span><span class="n">count</span><span class="p">])</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">variable_b_array</span><span class="p">[</span><span class="n">count</span><span class="p">]))</span>


    <span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;fold_change&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fc_list</span>
    <span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;t_statistic&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ttest_tval_list</span>
    <span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;p_value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ttest_pval_list</span>
    <span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;significance&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">signif_thres_list</span>

    <span class="k">del</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">val</span><span class="p">,</span> <span class="n">ttest_val</span><span class="p">,</span> <span class="n">ttest_tval_list</span><span class="p">,</span> <span class="n">ttest_pval_list</span><span class="p">,</span> <span class="n">fc_list</span><span class="p">,</span> <span class="n">signif_thres_list</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">normalize</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">del</span> <span class="p">(</span><span class="n">norm_a</span><span class="p">,</span> <span class="n">norm_b</span><span class="p">)</span>


<span class="c1">#%% Volcanoplot</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">19.0</span><span class="p">,</span><span class="mf">9.0</span><span class="p">))</span><span class="c1">#(27.0,3))</span>
    <span class="n">grid</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">0.0</span><span class="p">)</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">grid</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">])</span>

    <span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="kc">False</span><span class="p">:</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">:</span><span class="s1">&#39;red&#39;</span><span class="p">}</span>
    <span class="n">sc</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;fold_change&#39;</span><span class="p">],</span> <span class="n">y</span><span class="o">=</span><span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;p_value&#39;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;significance&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span><span class="n">colors</span><span class="p">[</span><span class="n">x</span><span class="p">]))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;major&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Log2 FC&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;-1*Log10 p-value&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">figure_title</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">variable</span> <span class="o">+</span> <span class="s2">&quot; - &quot;</span> <span class="o">+</span> <span class="n">figure_title</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">variable</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="p">[],</span><span class="n">y</span><span class="o">=</span><span class="p">[],</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;p-value &gt; </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">significance_threshold</span><span class="p">))</span> <span class="c1">#set empty scatterplot for legend</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="p">[],</span><span class="n">y</span><span class="o">=</span><span class="p">[],</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;p-value &lt; </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">significance_threshold</span><span class="p">))</span> <span class="c1">#set empty scatterplot for legend</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">trackgene_list</span> <span class="o">==</span> <span class="p">[]:</span>
        <span class="n">genenames_array</span> <span class="o">=</span> <span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;gene_names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">trackgene</span> <span class="ow">in</span> <span class="n">trackgene_list</span><span class="p">:</span>
            <span class="n">trackgene</span> <span class="o">=</span> <span class="n">trackgene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">trackgene</span> <span class="ow">in</span> <span class="n">genenames_array</span><span class="p">:</span>
                <span class="n">trackgene_index</span> <span class="o">=</span> <span class="n">tnread_gene_a</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">tnread_gene_a</span><span class="p">[</span><span class="s1">&#39;gene_names&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">trackgene</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">trackgene_annot</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">volcano_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">trackgene_index</span><span class="p">,:][</span><span class="s1">&#39;gene_names&#39;</span><span class="p">],</span> <span class="p">(</span><span class="n">volcano_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">trackgene_index</span><span class="p">,:][</span><span class="s1">&#39;fold_change&#39;</span><span class="p">],</span> <span class="n">volcano_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">trackgene_index</span><span class="p">,:][</span><span class="s1">&#39;p_value&#39;</span><span class="p">]),</span>
                            <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">boxstyle</span><span class="o">=</span><span class="s2">&quot;round&quot;</span><span class="p">,</span> <span class="n">fc</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">))</span>
                <span class="n">trackgene_annot</span><span class="o">.</span><span class="n">get_bbox_patch</span><span class="p">()</span><span class="o">.</span><span class="n">set_alpha</span><span class="p">(</span><span class="mf">0.6</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;WARNING: </span><span class="si">%s</span><span class="s1"> not found&#39;</span> <span class="o">%</span> <span class="n">trackgene</span><span class="p">)</span>
        <span class="k">del</span> <span class="p">(</span><span class="n">tnread_gene_a</span><span class="p">,</span> <span class="n">genenames_array</span><span class="p">)</span>


    <span class="n">names</span> <span class="o">=</span> <span class="n">volcano_df</span><span class="p">[</span><span class="s1">&#39;gene_names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">annot</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">xy</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="n">xytext</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span><span class="n">textcoords</span><span class="o">=</span><span class="s2">&quot;offset points&quot;</span><span class="p">,</span>
                        <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">boxstyle</span><span class="o">=</span><span class="s2">&quot;round&quot;</span><span class="p">,</span> <span class="n">fc</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">),</span>
                        <span class="n">arrowprops</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">arrowstyle</span><span class="o">=</span><span class="s2">&quot;-&gt;&quot;</span><span class="p">))</span>
    <span class="n">annot</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">update_annot</span><span class="p">(</span><span class="n">ind</span><span class="p">):</span>

        <span class="n">pos</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get_offsets</span><span class="p">()[</span><span class="n">ind</span><span class="p">[</span><span class="s2">&quot;ind&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
        <span class="n">annot</span><span class="o">.</span><span class="n">xy</span> <span class="o">=</span> <span class="n">pos</span>
        <span class="c1"># text = &quot;{}, {}&quot;.format(&quot; &quot;.join(list(map(str,ind[&quot;ind&quot;]))), </span>
        <span class="c1">#                         &quot; &quot;.join([names[n] for n in ind[&quot;ind&quot;]]))</span>
        <span class="n">text</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">names</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">ind</span><span class="p">[</span><span class="s2">&quot;ind&quot;</span><span class="p">]]))</span>
        <span class="n">annot</span><span class="o">.</span><span class="n">set_text</span><span class="p">(</span><span class="n">text</span><span class="p">)</span>
        <span class="c1"># annot.get_bbox_patch().set_facecolor(cmap(norm(c[ind[&quot;ind&quot;][0]])))</span>
        <span class="c1"># annot.get_bbox_patch().set_alpha(0.4)</span>


    <span class="k">def</span> <span class="nf">hover</span><span class="p">(</span><span class="n">event</span><span class="p">):</span>
        <span class="n">vis</span> <span class="o">=</span> <span class="n">annot</span><span class="o">.</span><span class="n">get_visible</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">event</span><span class="o">.</span><span class="n">inaxes</span> <span class="o">==</span> <span class="n">ax</span><span class="p">:</span>
            <span class="n">cont</span><span class="p">,</span> <span class="n">ind</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">event</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">cont</span><span class="p">:</span>
                <span class="n">update_annot</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>
                <span class="n">annot</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">draw_idle</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">vis</span><span class="p">:</span>
                    <span class="n">annot</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                    <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">draw_idle</span><span class="p">()</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">mpl_connect</span><span class="p">(</span><span class="s2">&quot;motion_notify_event&quot;</span><span class="p">,</span> <span class="n">hover</span><span class="p">)</span>


<span class="c1">#%% return function</span>
    <span class="k">return</span><span class="p">(</span><span class="n">volcano_df</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="section" id="how-to-use-the-code-in-the-same-script">
<h2>How to use the code in the same script<a class="headerlink" href="#how-to-use-the-code-in-the-same-script" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">volcano_df</span> <span class="o">=</span> <span class="n">volcano</span><span class="p">(</span><span class="n">path_a</span><span class="o">=</span><span class="n">path_a</span><span class="p">,</span> <span class="n">filelist_a</span><span class="o">=</span><span class="n">filelist_a</span><span class="p">,</span>
            <span class="n">path_b</span><span class="o">=</span><span class="n">path_b</span><span class="p">,</span> <span class="n">filelist_b</span><span class="o">=</span><span class="n">filelist_b</span><span class="p">,</span>
            <span class="n">variable</span><span class="o">=</span><span class="n">variable</span><span class="p">,</span>
            <span class="n">significance_threshold</span><span class="o">=</span><span class="n">significance_threshold</span><span class="p">,</span>
            <span class="n">normalize</span><span class="o">=</span><span class="n">normalize</span><span class="p">,</span>
            <span class="n">trackgene_list</span><span class="o">=</span><span class="n">trackgene_list</span><span class="p">,</span>
            <span class="n">figure_title</span><span class="o">=</span><span class="n">figure_title</span><span class="p">)</span>

</pre></div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./02-postprocessing-scripts"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="04-scatterplot_genes.html" title="previous page">Create a scatterplot for all genes and all essential genes.</a>
    <a class='right-next' id="next-link" href="06-genomicfeatures_dataframe.html" title="next page">Searching for genomic features  of transposition events</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Gregory van Beek , Maurits Kok, Leila Iñigo de la Cruz<br/>
        
            &copy; Copyright Apache 2.0.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.1c5a1a01449ed65a7b51.js"></script>

  
  </body>
</html>