{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python Translation of the Matlab code from Kornmann Lab \n",
    "\n",
    "```{note}\n",
    "\n",
    "This is a test script file for translating the Matlab code from the Kornmann lab to Python.\n",
    "\n",
    "To save  numpy array in text format:\n",
    "    np.savetxt(os.path.join(dirname,'tncoordinatescopy_python.txt'),tncoordinatescopy_array,delimiter=',',fmt='%i')\n",
    "\"\"\"\n",
    "```\n",
    "\n",
    "``` python\n",
    "import os, sys\n",
    "import numpy as np\n",
    "import pysam\n",
    "import timeit\n",
    "\n",
    "dirname = os.path.dirname(os.path.abspath('__file__'))\n",
    "sys.path.insert(1,os.path.join(dirname,'python_modules'))\n",
    "from chromosome_and_gene_positions import chromosomename_roman_to_arabic, gene_position\n",
    "from gene_names import gene_aliases\n",
    "from essential_genes_names import list_known_essentials\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: START TIMER\n",
    "\n",
    "timer_start_complete = timeit.default_timer()\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: LOADING FILES\n",
    "\n",
    "path = os.path.join('/home', 'gregoryvanbeek', 'Documents', 'data_processing')\n",
    "# filename = os.path.join('E-MTAB-4885.WT2.bam')\n",
    "filename = os.path.join('E-MTAB-4885.WT1.bam')\n",
    "\n",
    "file = os.path.join(path,filename)\n",
    "print('Running path: ', file)\n",
    "\n",
    "if os.path.isfile(file):\n",
    "    print('File exists.')\n",
    "elif os.path.exists(file):\n",
    "    print('File does not exist, but path does exists.')\n",
    "else:\n",
    "    print('Path does not exist.')\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: READ BAM FILE\n",
    "\n",
    "bamfile = pysam.AlignmentFile(file, 'rb')\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: GET NAMES OF ALL CHROMOSOMES AS STORED IN THE BAM FILE\n",
    "\n",
    "ref_tid_dict = {} # 'I' | 0\n",
    "ref_name_list = []\n",
    "for i in range(17):\n",
    "    ref_name = bamfile.get_reference_name(i)\n",
    "    ref_tid_dict[ref_name] = bamfile.get_tid(ref_name)\n",
    "    ref_name_list.append(ref_name)\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: GET SEQUENCE LENGTHS OF ALL CHROMOSOMES\n",
    "\n",
    "chr_length_dict = {} # 'I' | 230218\n",
    "chr_summedlength_dict = {}\n",
    "ll = 0\n",
    "for key in ref_tid_dict:\n",
    "    ref_length = bamfile.get_reference_length(key)\n",
    "    chr_length_dict[key] = ref_length\n",
    "    chr_summedlength_dict[key] = ll\n",
    "    ll += ref_length\n",
    "del (ref_length, ll)\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: GET NUMBER OF MAPPED, UNMAPPED AND TOTAL AMOUNT OF READS PER CHROMOSOME\n",
    "\n",
    "chr_mappedreads_dict = {} # 'I' | [mapped, unmapped, total reads]\n",
    "total_reads = 0\n",
    "for i in range(17):\n",
    "    stats = bamfile.get_index_statistics()[i]\n",
    "    chr_mappedreads_dict[stats[0]] = [stats[1], stats[2], stats[3]]\n",
    "    total_reads += stats[3]\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: GET ALL READS WITHIN A SPECIFIED GENOMIC REGION\n",
    "\n",
    "# readnumb_array = [] #np.array([], dtype=int)\n",
    "tnnumber_dict = {}\n",
    "\n",
    "ll = 0 #Number of unique insertions in entire genome\n",
    "# temp = ['I','II']\n",
    "for kk in ref_name_list: # 'kk' is chromosome number in roman numerals\n",
    "    read_counter = 0\n",
    "    timer_start = timeit.default_timer()\n",
    "    \n",
    "    N_reads_kk = chr_mappedreads_dict[kk][2]\n",
    "    start_array = np.empty(shape=(N_reads_kk), dtype=int)\n",
    "    flag_array = np.empty(shape=(N_reads_kk), dtype=int)\n",
    "    readlength_array = np.empty(shape=(N_reads_kk), dtype=int)\n",
    "\n",
    "    print('Getting reads for chromosme ', kk ,' ...')\n",
    "    for reads in bamfile.fetch(kk, 0, chr_length_dict[kk]):\n",
    "        read = str(reads).split('\\t')\n",
    "        \n",
    "        start_array[read_counter] = int(read[3]) + 1\n",
    "        flag_array[read_counter] = int(read[1])\n",
    "        readlength_array[read_counter] = int(len(read[9]))\n",
    "\n",
    "        read_counter += 1\n",
    "\n",
    "\n",
    "##% CORRECT STARTING POSITION FOR READS WITH REVERSED ORIENTATION\n",
    "    flag0coor_array = np.where(flag_array==0) #coordinates reads 5' -> 3'\n",
    "    flag16coor_array = np.where(flag_array==16) # coordinates reads 3' -> 5'\n",
    "\n",
    "    startdirect_array = start_array[flag0coor_array]\n",
    "    flagdirect_array = flag_array[flag0coor_array]\n",
    "\n",
    "    startindirect_array = start_array[flag16coor_array] + readlength_array[flag16coor_array]\n",
    "    flagindirect_array = flag_array[flag16coor_array]\n",
    "\n",
    "    start2_array = np.concatenate((startdirect_array, startindirect_array), axis=0)\n",
    "    flag2_array = np.concatenate((flagdirect_array, flagindirect_array), axis=0)\n",
    "    \n",
    "    del flag0coor_array, flag16coor_array, startdirect_array, flagdirect_array, startindirect_array, flagindirect_array\n",
    "    \n",
    "    \n",
    "    start2_sortindices = start2_array.argsort(kind='mergesort') #use mergesort for stable sorting\n",
    "    start2_array = start2_array[start2_sortindices]\n",
    "    flag2_array = flag2_array[start2_sortindices]\n",
    "    \n",
    "    del start2_sortindices\n",
    "    \n",
    "    \n",
    "##% CREATE ARRAY OF START POSITION AND FLAGS OF ALL READS IN GENOME\n",
    "    ref_tid_kk = int(ref_tid_dict[kk]+1)\n",
    "    if ll == 0:\n",
    "        tncoordinates_array = np.array([])\n",
    "    \n",
    "    mm = 0 # Number of unique reads per insertion\n",
    "    jj = 1 # Number of unique reads in current chromosome (Number of transposons in current chromosome)\n",
    "    temp_counter = 0\n",
    "    for ii in range(1,len(start2_array)):\n",
    "        if abs(start2_array[ii]-start2_array[ii-1]) <= 2 and flag2_array[ii] == flag2_array[ii-1]:\n",
    "            mm += 1\n",
    "            temp_counter += 1\n",
    "        else:\n",
    "            avg_start_pos = abs(round(np.mean(start2_array[ii-mm-1 : ii])))\n",
    "            if tncoordinates_array.size == 0:\n",
    "                tncoordinates_array = np.array([ref_tid_kk, int(avg_start_pos), int(flag2_array[ii-1])])\n",
    "                readnumb_list = [mm+1]\n",
    "            else:\n",
    "                tncoordinates_array = np.vstack((tncoordinates_array, [ref_tid_kk, int(avg_start_pos), int(flag2_array[ii-1])]))    \n",
    "                readnumb_list.append(mm+1)\n",
    "            mm = 0\n",
    "            jj += 1\n",
    "            ll += 1\n",
    "        \n",
    "        if ii == len(start2_array) - 1: #include last read\n",
    "            avg_start_pos = abs(round(np.mean(start2_array[ii-mm-1 : ii])))\n",
    "            tncoordinates_array = np.vstack((tncoordinates_array, [ref_tid_kk, int(avg_start_pos), int(flag2_array[ii-1])]))\n",
    "            readnumb_list.append(mm+1)\n",
    "\n",
    "    tnnumber_dict[kk] = jj\n",
    "    \n",
    "    del jj, start_array, flag_array, readlength_array, flag2_array, start2_array\n",
    "\n",
    "    timer_stop = timeit.default_timer()\n",
    "    print('Loop over reads chromosome ', kk, ' complete. Time = ', timer_stop-timer_start, 'seconds')\n",
    "\n",
    "readnumb_array = np.array(readnumb_list)\n",
    "del readnumb_list\n",
    "\n",
    "tncoordinatescopy_array = np.array(tncoordinates_array, copy=True)\n",
    "# np.savetxt(os.path.join(dirname,'tncoordinates_python.txt'),tncoordinates_array,delimiter=',',fmt='%i') #!!!\n",
    "# np.savetxt(os.path.join(dirname,'readnumb_python.txt'),readnumb_array,delimiter=',',fmt='%i') #!!!\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: GET LIST OF ALL GENES AND ALL ESSENTIAL GENES\n",
    "\n",
    "\n",
    "files_path = os.path.join(dirname,'..','data_files')\n",
    "chromosomenames_gff_dict = chromosomename_roman_to_arabic()[0] #These names should correspond to the chromosome names in the gff files\n",
    "\n",
    "# GET POSITION GENES\n",
    "gff_path = os.path.join(files_path,'Saccharomyces_cerevisiae.R64-1-1.99.gff3')\n",
    "genecoordinates_dict = gene_position(gff_path) #contains all genes, essential and nonessential\n",
    "# gff_gene_name, gff_gene_chr, gff_gene_start, gff_gene_end = gene_position(gff_path, get_dict=False)[0:4]\n",
    "\n",
    "# GET ALL ANNOTATED ESSENTIAL GENES\n",
    "essential_path1 = os.path.join(files_path,'Cervisiae_EssentialGenes_List_1.txt')\n",
    "essential_path2 = os.path.join(files_path,'Cervisiae_EssentialGenes_List_2.txt')\n",
    "essentialnames_list = list_known_essentials([essential_path1, essential_path2])\n",
    "\n",
    "# GET ALIASES OF ALL GENES\n",
    "names_path = os.path.join(files_path,'Yeast_Protein_Names.txt')\n",
    "aliases_designation_dict = gene_aliases(names_path)[0]\n",
    "\n",
    "# FOR ALL GENE IN GENECOORDINATES_DICT, CHECK IF THEY ARE ANNOTATED AS ESSENTIAL\n",
    "essentialcoordinates_dict = {}\n",
    "gene_counter = 0\n",
    "for gene in genecoordinates_dict:\n",
    "# for gene in gff_gene_name:\n",
    "    if gene in essentialnames_list:\n",
    "        essentialcoordinates_dict[gene] = genecoordinates_dict.get(gene).copy()\n",
    "        # essentialcoordinates_dict[gene] = [gff_gene_chr[gene_counter], gff_gene_start[gene_counter], gff_gene_end[gene_counter]]\n",
    "    else:\n",
    "        gene_aliases_list = []\n",
    "        for key, val in aliases_designation_dict.items():\n",
    "            if gene == key or gene in val: #if gene occurs as key or in the values list in aliases_designation_dict, put all its aliases in a single list.\n",
    "                gene_aliases_list.append(key)\n",
    "                for aliases in aliases_designation_dict.get(key):\n",
    "                    gene_aliases_list.append(aliases)\n",
    "\n",
    "        for gene_alias in gene_aliases_list:\n",
    "            if gene_alias in essentialnames_list:\n",
    "                essentialcoordinates_dict[gene_alias] = genecoordinates_dict.get(gene).copy()\n",
    "                # essentialcoordinates_dict[gene_alias] = [gff_gene_chr[gene_counter], gff_gene_start[gene_counter], gff_gene_end[gene_counter]]\n",
    "                break\n",
    "    gene_counter += 1\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: CONCATENATE ALL CHROMOSOMES\n",
    "\n",
    "ll = 0\n",
    "for ii in range(1,len(ref_name_list)):\n",
    "    ll += chr_length_dict[ref_name_list[ii-1]]\n",
    "    aa = np.where(tncoordinatescopy_array[:,0] == ii + 1)\n",
    "    tncoordinatescopy_array[aa,1] = tncoordinatescopy_array[aa,1] + ll\n",
    "\n",
    "for key in genecoordinates_dict:\n",
    "    genecoordinates_dict[key][1] = genecoordinates_dict.get(key)[1] + chr_summedlength_dict.get(genecoordinates_dict.get(key)[0])\n",
    "    genecoordinates_dict[key][2] = genecoordinates_dict.get(key)[2] + chr_summedlength_dict.get(genecoordinates_dict.get(key)[0])\n",
    "\n",
    "for key in essentialcoordinates_dict:\n",
    "    essentialcoordinates_dict[key][1] = essentialcoordinates_dict.get(key)[1] + chr_summedlength_dict.get(essentialcoordinates_dict.get(key)[0])\n",
    "    essentialcoordinates_dict[key][2] = essentialcoordinates_dict.get(key)[2] + chr_summedlength_dict.get(essentialcoordinates_dict.get(key)[0])\n",
    "\n",
    "# test = []\n",
    "# for key in genecoordinates_dict:\n",
    "#     test.append(genecoordinates_dict.get(key)[1:3])\n",
    "#np.savetxt(os.path.join(dirname,'essentialcoordinates_python.txt'),test,delimiter=',',fmt='%i')\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: GET NUMBER OF TRANSPOSONS AND READS PER GENE\n",
    "\n",
    "tnpergene_dict = {}\n",
    "readpergene_dict = {}\n",
    "readpergenecrude_dict = {}\n",
    "for gene in genecoordinates_dict:\n",
    "    xx = np.where(np.logical_and(tncoordinatescopy_array[:,1] >= genecoordinates_dict.get(gene)[1], tncoordinatescopy_array[:,1] <= genecoordinates_dict.get(gene)[2]))\n",
    "    tnpergene_dict[gene] = np.size(xx)\n",
    "    readpergene_dict[gene] = sum(readnumb_array[xx]) - max(readnumb_array[xx], default=0)\n",
    "    readpergenecrude_dict[gene] = sum(readnumb_array[xx])\n",
    "\n",
    "\n",
    "tnperessential_dict = {}\n",
    "readperessential_dict = {}\n",
    "readperessentialcrude_dict = {}\n",
    "for gene in essentialcoordinates_dict:\n",
    "    xx = np.where(np.logical_and(tncoordinatescopy_array[:,1] >= essentialcoordinates_dict.get(gene)[1], tncoordinatescopy_array[:,1] <= essentialcoordinates_dict.get(gene)[2]))\n",
    "    readperessential_dict[gene] = sum(readnumb_array[xx]) - max(readnumb_array[xx], default=0)\n",
    "    readperessentialcrude_dict[gene] = sum(readnumb_array[xx])\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: CREATE BED FILE\n",
    "\n",
    "bedfile = file+'.bed'\n",
    "\n",
    "with open(bedfile, 'w') as f:\n",
    "    f.write('track name=' + filename + ' useScore=1\\n')\n",
    "    coordinates_counter = 0\n",
    "    for tn in tncoordinates_array:\n",
    "        refname = [key for key, val in ref_tid_dict.items() if val == tn[0] - 1][0]\n",
    "        if refname == 'Mito':\n",
    "            refname = 'M'\n",
    "        f.write('chr' + refname + ' ' + str(tn[1]) + ' ' + str(tn[1] + 1) + ' . ' + str(100+readnumb_array[coordinates_counter]*20) + '\\n')\n",
    "        coordinates_counter += 1\n",
    "\n",
    "\n",
    "#CHECK NUMBER OF INSERTIONS FOR EACH CHROMOSOME (NOTE, THIS IS SLOW)\n",
    "# test_list = []\n",
    "# for i in range(1,18):\n",
    "#     test_counter = 0\n",
    "#     for tn in tncoordinates_array:\n",
    "#         if tn[0] == i:\n",
    "#             test_counter += 1\n",
    "#     test_list.append([i,test_counter])\n",
    "# test_list\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: CREATE TEXT FILE WITH TRANSPOSONS AND READS PER GENE\n",
    "\n",
    "pergenefile = file+'_pergene.txt'\n",
    "\n",
    "with open(pergenefile, 'w') as f:\n",
    "    f.write('Gene name\\tNumber of transposons per gene\\tNumber of reads per gene\\n')\n",
    "    for gene in tnpergene_dict:\n",
    "        tnpergene = tnpergene_dict[gene]\n",
    "        readpergene = readpergene_dict[gene]\n",
    "        f.write(gene + '\\t' + str(tnpergene) + '\\t' + str(readpergene) + '\\n')\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: CREATE WIG FILE\n",
    "\n",
    "wigfile = file+'.wig'\n",
    "\n",
    "readnumbwig_array = readnumb_array.copy()\n",
    "\n",
    "unique_index_array = np.array([], dtype=int) #=cc\n",
    "N_uniques_perchr_list = []\n",
    "ll = 0\n",
    "for kk in ref_name_list:\n",
    "    index = np.where(tncoordinates_array[:,0] == int(ref_tid_dict[kk]+1)) #get indices for current chromosome.\n",
    "    unique_index = np.unique(tncoordinates_array[index][:,1], return_index=True)[1] #get all insertion locations (in tncoordinates, all rows, column 1)\n",
    "\n",
    "    unique_index_array = np.append(unique_index_array, (unique_index+ll), axis=0)\n",
    "    \n",
    "    if ll == 0:\n",
    "        ll += np.count_nonzero(tncoordinates_array[:,0] == int(ref_tid_dict[kk]+1))# + 1\n",
    "    else:\n",
    "        ll += np.count_nonzero(tncoordinates_array[:,0] == int(ref_tid_dict[kk]+1))\n",
    "    N_uniques_perchr_list.append(ll) #total amount unique indices found untill current chromosome\n",
    "\n",
    "# np.savetxt(os.path.join(dirname,'uniqueindex_python.txt'),unique_index_array,delimiter=',',fmt='%i')\n",
    "\n",
    "duplicate_list = [] #=dd\n",
    "ll = 0\n",
    "index_last_unique_previous_chromosome = 0\n",
    "for ii in N_uniques_perchr_list:\n",
    "    index_last_unique = np.where(unique_index_array <= ii)[0][-1]\n",
    "    for jj in range(ll,ii):\n",
    "        if int(jj) not in unique_index_array[index_last_unique_previous_chromosome:index_last_unique]:\n",
    "            if ll == 0:\n",
    "                duplicate_list.append(jj)\n",
    "            else:\n",
    "                duplicate_list.append(jj)#-1)\n",
    "    index_last_unique_previous_chromosome = index_last_unique\n",
    "    ll = ii\n",
    "# np.savetxt(os.path.join(dirname,'duplicateindex_python.txt'),duplicate_list,delimiter=',',fmt='%i')\n",
    "# testposition_dict = {}\n",
    "# testorientation_dict = {}\n",
    "# for ii in duplicate_list:\n",
    "#     testposition_dict[ii] = tncoordinates_array[ii][1] - tncoordinates_array[ii-1][1]\n",
    "#     testorientation_dict[ii] = tncoordinates_array[ii][2] - tncoordinates_array[ii-1][2]\n",
    "\n",
    "\n",
    "#SUM READNUMB VALUES AT INDEX IN DUPLICATE_LIST AND DUPLICATE_LIST-1    \n",
    "for ii in duplicate_list:\n",
    "    readnumbwig_array[ii-1] = readnumbwig_array[ii-1] + readnumbwig_array[ii]\n",
    "\n",
    "readnumbwig_duplicatesremoved_array = np.delete(readnumbwig_array, duplicate_list, axis=0)\n",
    "tncoordinateswig_duplicatesremoved_array = np.delete(tncoordinates_array, duplicate_list, axis=0)\n",
    "\n",
    "\n",
    "# del duplicate_list, unique_index_array, N_uniques_perchr_list\n",
    "\n",
    "\n",
    "with  open(wigfile, 'w') as f:\n",
    "    f.write('track type=wiggle_0 ,maxheightPixels=60 name='+filename+'\\n')\n",
    "    for kk in ref_name_list:\n",
    "        f.write('VariableStep chrom=chr' + kk + '\\n')\n",
    "        \n",
    "        index = np.where(tncoordinateswig_duplicatesremoved_array[:,0] == int(ref_tid_dict[kk]+1)) #get indices for current chromosome.\n",
    "        for ii in index[0]:\n",
    "            f.write(str(tncoordinateswig_duplicatesremoved_array[ii][1]) + ' ' + str(readnumbwig_duplicatesremoved_array[ii]) + '\\n')\n",
    "```\n",
    "\n",
    "```python\n",
    ":title: END TIMER\n",
    "\n",
    "timer_stop_complete = timeit.default_timer()\n",
    "print('Script took %.2f seconds to run' %(timer_stop_complete - timer_start_complete))\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".md",
    "format_name": "myst",
    "format_version": "0.8",
    "jupytext_version": "1.4.2"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  },
  "source_map": [
   11
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 4
}