
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>3. p-value according a T-student distribution &#8212; SATAY DATA ANALYSIS LAAN-LAB TU DELFT</title>
    
  <link rel="stylesheet" href="../_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx_tabs/semantic-ui-2.4.1/segment.min.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx_tabs/semantic-ui-2.4.1/menu.min.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx_tabs/semantic-ui-2.4.1/tab.min.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx_tabs/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/Greg.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="2. From reads and insertions to essentiality prediction using a Regression Model" href="from-reads-to-essential-genes.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  <img src="../_static/Greg.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">SATAY DATA ANALYSIS LAAN-LAB TU DELFT</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../Introduction.html">
   Status of the project
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Get started
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../Intro/flowchart.html">
   1. Flowchart of the pipeline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Intro/satay_analysis_notes.html">
   2. All what you need to know from SATAY
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Intro/data-analysis-workflow.html">
   3. Data analysis steps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Intro/Installation_Guide_SATAY_Analysis_Software.html">
   4. Software installation guide for whole genome sequencing and transposon mapping
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Bash GUI
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../bash-GUI/how-to-use-the-bash-gui.html">
   1. Main steps of the BASH GUI
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Data Stats
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../data-stats/transposonmapping_satay.html">
   1. Python Translation of the Matlab code from Kornmann Lab
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../data-stats/statistics_pergene.html">
   2. Statistics per gene
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../data-stats/statistics_perchromosome_Compare.html">
   3. Comparison of the statistics per chromosome
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Postprocessing scripts
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="gene_reads.html">
   1. Gene reads script
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="from-reads-to-essential-genes.html">
   2. From reads and insertions to essentiality prediction using a Regression Model
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   3. p-value according a T-student distribution
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/postprocessing-scripts/volcano_plots_reproducing-Benoit-paper.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/leilaicruz/SATAY-jupyter-book"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/leilaicruz/SATAY-jupyter-book/issues/new?title=Issue%20on%20page%20%2Fpostprocessing-scripts/volcano_plots_reproducing-Benoit-paper.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org//v2/gh/leilaicruz/SATAY-jupyter-book/gh-pages?urlpath=lab/tree/postprocessing-scripts/volcano_plots_reproducing-Benoit-paper.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="o">%</span><span class="k">matplotlib</span> inline 
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">from</span> <span class="nn">bioinfokit</span> <span class="kn">import</span> <span class="n">analys</span><span class="p">,</span> <span class="n">visuz</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">script_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="s1">&#39;__file__&#39;</span><span class="p">)</span> <span class="c1">#&lt;-- absolute dir the script is in</span>
<span class="c1"># rel_path_data_wt=&quot;WT_trimmed-sorted-bam.txt&quot;</span>
<span class="n">rel_path_data_wt</span><span class="o">=</span><span class="s2">&quot;benoit-Processed-dataset-WildType1_pergene.txt&quot;</span>
<span class="n">rel_path_data_mutant</span><span class="o">=</span><span class="s2">&quot;benoit-Processed-dataset-Dpl1del_pergene.txt&quot;</span>
<span class="c1"># rel_path_data_mutant=&quot;dDpl1Kan-sorted-bam.txt&quot;</span>


<span class="n">abs_path_data_wt</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">script_dir</span><span class="p">,</span> <span class="n">rel_path_data_wt</span><span class="p">)</span>
<span class="n">abs_path_data_mutant</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">script_dir</span><span class="p">,</span> <span class="n">rel_path_data_mutant</span><span class="p">)</span>
<span class="n">data_wt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">abs_path_data_wt</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">data_mutant</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">abs_path_data_mutant</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># data_mutant.columns=[&#39;insertions&#39;,&#39;reads&#39;]</span>
<span class="c1"># data_wt.columns=[&#39;insertions&#39;,&#39;reads&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_wt</span><span class="p">[</span><span class="s1">&#39;number_of_read_pergene&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_wt</span><span class="p">[</span><span class="s1">&#39;number_of_transposon_per_gene&#39;</span><span class="p">]</span>
<span class="n">data_wt</span><span class="p">[</span><span class="s1">&#39;number_of_transposon_per_gene&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_wt</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span>
<span class="n">data_wt</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_wt</span><span class="o">.</span><span class="n">index</span>
<span class="n">data_wt</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">data_wt</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_mutant</span><span class="p">[</span><span class="s1">&#39;number_of_read_pergene&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_mutant</span><span class="p">[</span><span class="s1">&#39;number_of_transposon_per_gene&#39;</span><span class="p">]</span>
<span class="n">data_mutant</span><span class="p">[</span><span class="s1">&#39;number_of_transposon_per_gene&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_mutant</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span>
<span class="n">data_mutant</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_mutant</span><span class="o">.</span><span class="n">index</span>
<span class="n">data_mutant</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">data_mutant</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_mutant</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;geneid&#39;</span><span class="p">,</span><span class="s1">&#39;insertions&#39;</span><span class="p">,</span><span class="s1">&#39;reads&#39;</span><span class="p">]</span>
<span class="n">data_wt</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;geneid&#39;</span><span class="p">,</span><span class="s1">&#39;insertions&#39;</span><span class="p">,</span><span class="s1">&#39;reads&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">total_tn_mutant</span><span class="o">=</span><span class="n">data_mutant</span><span class="o">.</span><span class="n">insertions</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="n">total_tn_wt</span><span class="o">=</span><span class="n">data_wt</span><span class="o">.</span><span class="n">insertions</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_mutant_normalized2tn</span><span class="o">=</span><span class="n">data_mutant</span><span class="o">.</span><span class="n">insertions</span><span class="o">/</span><span class="n">total_tn_mutant</span>
<span class="n">data_wt_normalized2tn</span><span class="o">=</span><span class="n">data_wt</span><span class="o">.</span><span class="n">insertions</span><span class="o">/</span><span class="n">total_tn_wt</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fold_change_per_gene</span><span class="o">=</span><span class="n">data_mutant_normalized2tn</span><span class="o">/</span><span class="n">data_wt_normalized2tn</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data2volcano</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;geneid&#39;</span><span class="p">,</span><span class="s1">&#39;values-exp&#39;</span><span class="p">,</span><span class="s1">&#39;values-ref&#39;</span><span class="p">,</span><span class="s1">&#39;log2FC&#39;</span><span class="p">,</span><span class="s1">&#39;p-value&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fold_change_per_gene</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0       1.016231
1       2.328864
2       0.465773
3       1.397318
4       1.019665
          ...   
6598    1.571983
6599    1.130184
6600    0.877205
6601    0.613973
6602    0.931546
Name: insertions, Length: 6603, dtype: float64
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;geneid&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_wt</span><span class="p">[</span><span class="s1">&#39;geneid&#39;</span><span class="p">]</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;values-exp&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_mutant_normalized2tn</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;values-ref&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data_wt_normalized2tn</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">fold_change_per_gene</span><span class="p">)</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;values-ref&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;values-exp&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">to_replace</span><span class="o">=</span><span class="s1">&#39;-inf&#39;</span><span class="p">,</span><span class="n">value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="p-value-according-a-t-student-distribution">
<h1><span class="section-number">3. </span>p-value according a T-student distribution<a class="headerlink" href="#p-value-according-a-t-student-distribution" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://towardsdatascience.com/inferential-statistics-series-t-test-using-numpy-2718f8f9bf2f">https://towardsdatascience.com/inferential-statistics-series-t-test-using-numpy-2718f8f9bf2f</a></p></li>
<li><p><a class="reference external" href="https://www.reneshbedre.com/blog/volcano.html">https://www.reneshbedre.com/blog/volcano.html</a></p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define 2 random distributions</span>
<span class="c1">#Sample Size</span>
<span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data2volcano</span><span class="p">)</span>
<span class="c1">#Gaussian distributed data with mean = 2 and var = 1</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;values-exp&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="c1">#Gaussian distributed data with with mean = 0 and var = 1</span>
<span class="n">b</span> <span class="o">=</span> <span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;values-ref&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>


<span class="c1">## Calculate the Standard Deviation</span>
<span class="c1">#Calculate the variance to get the standard deviation</span>

<span class="c1">#For unbiased max likelihood estimate we have to divide the var by N-1, and therefore the parameter ddof = 1</span>
<span class="n">var_a</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">ddof</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">var_b</span> <span class="o">=</span> <span class="n">b</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">ddof</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1">#std deviation</span>
<span class="n">s</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">((</span><span class="n">var_a</span> <span class="o">+</span> <span class="n">var_b</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>

<span class="c1">## Calculate the t-statistics</span>
<span class="n">t</span> <span class="o">=</span> <span class="p">(</span><span class="n">a</span> <span class="o">-</span> <span class="n">b</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">s</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mi">2</span><span class="o">/</span><span class="n">N</span><span class="p">))</span>


<span class="c1">## Compare with the critical t-value</span>
<span class="c1">#Degrees of freedom</span>
<span class="n">df</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">N</span> <span class="o">-</span> <span class="mi">2</span>

<span class="c1">#p-value after comparison with the t </span>
<span class="n">p</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">stats</span><span class="o">.</span><span class="n">t</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">t</span><span class="p">,</span><span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="p">)</span>


<span class="c1"># print(&quot;t = &quot; + str(t))</span>
<span class="c1"># print(&quot;p = &quot; + str(2*p))</span>
<span class="c1">### You can see that after comparing the t statistic with the critical t value (computed internally) we get a good p value of 0.0005 and thus we reject the null hypothesis and thus it proves that the mean of the two distributions are different and statistically significant.</span>


<span class="c1">## Cross Checking with the internal scipy function</span>
<span class="n">t2</span><span class="p">,</span> <span class="n">p2</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span><span class="n">a</span><span class="p">,</span><span class="n">b</span><span class="p">)</span>
<span class="c1"># print(&quot;t = &quot; + str(t2))</span>
<span class="c1"># print(&quot;p = &quot; + str(p2))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">p</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;ttest&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">t</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;ttest&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(array([3.000e+00, 1.400e+01, 8.400e+01, 1.324e+03, 4.880e+03, 2.710e+02,
        1.300e+01, 5.000e+00, 1.000e+00, 2.000e+00]),
 array([-141.63018633, -108.25807165,  -74.88595696,  -41.51384228,
          -8.14172759,   25.23038709,   58.60250178,   91.97461646,
         125.34673115,  158.71884584,  192.09096052]),
 &lt;a list of 10 Patch objects&gt;)
</pre></div>
</div>
<img alt="../_images/volcano_plots_reproducing-Benoit-paper_15_1.png" src="../_images/volcano_plots_reproducing-Benoit-paper_15_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">inf</span><span class="o">=</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">][</span><span class="mi">6599</span><span class="p">]</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">inf</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;log2FC&#39;</span><span class="p">])</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">data2volcano</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visuz</span><span class="o">.</span><span class="n">gene_exp</span><span class="o">.</span><span class="n">volcano</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">data2volcano</span><span class="p">,</span> <span class="n">lfc</span><span class="o">=</span><span class="s1">&#39;log2FC&#39;</span><span class="p">,</span> <span class="n">pv</span><span class="o">=</span><span class="s1">&#39;p-value&#39;</span><span class="p">,</span><span class="n">pv_thr</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span><span class="n">geneid</span><span class="o">=</span><span class="s1">&#39;geneid&#39;</span><span class="p">,</span><span class="n">genenames</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;CDC24&#39;</span><span class="p">,</span><span class="s1">&#39;CDC42&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visuz</span><span class="o">.</span><span class="n">gene_exp</span><span class="o">.</span><span class="n">volcano</span><span class="p">(</span><span class="n">df</span><span class="o">=</span><span class="n">data2volcano</span><span class="p">,</span> <span class="n">lfc</span><span class="o">=</span><span class="s1">&#39;log2FC&#39;</span><span class="p">,</span> <span class="n">pv</span><span class="o">=</span><span class="s1">&#39;p-value&#39;</span><span class="p">,</span> <span class="n">plotlegend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">legendpos</span><span class="o">=</span><span class="s1">&#39;upper right&#39;</span><span class="p">,</span> 
    <span class="n">legendanchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.46</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">pv_thr</span><span class="o">=</span><span class="p">[</span><span class="mf">0.99</span><span class="p">,</span><span class="mf">0.99</span><span class="p">],</span><span class="n">sign_line</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python37064bitbaseconda6c8618fc0f0945779f8ded01abbb1089",
            path: "./postprocessing-scripts"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python37064bitbaseconda6c8618fc0f0945779f8ded01abbb1089'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="from-reads-to-essential-genes.html" title="previous page"><span class="section-number">2. </span>From reads and insertions to essentiality prediction using a Regression Model</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Gregory van Beek<br/>
        
            &copy; Copyright GPL License 3.0.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>